<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<title>Snakemake for Bioinformatics: All in One View</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<link rel="stylesheet" type="text/css" href="assets/styles.css">
<script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="manifest" href="site.webmanifest">
<link rel="mask-icon" href="safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" content="#ffffff">
</head>
<body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav incubator"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="Carpentries Incubator" src="assets/images/incubator-logo.svg"><abbr class="badge badge-light" title="This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team." style="background-color: #001483; border-radius: 5px">
          <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#polishing-beta-stage" class="external-link alert-link" style="color: #FFF7F1">
            <i aria-hidden="true" class="icon" data-feather="alert-circle" style="border-radius: 5px"></i>
            Beta
          </a>
          <span class="visually-hidden">This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.</span>
        </abbr>
        
      </div>
    </div>
    <div class="selector-container">
      
      
      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1">
<li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/aio.html';">Instructor View</button></li>
        </ul>
</div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo-sm.svg">
</div>
    <div class="lesson-title-md">
      Snakemake for Bioinformatics
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0">
<li class="nav-item">
          <span class="lesson-title">
            Snakemake for Bioinformatics
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown">
<li><a class="dropdown-item" href="discuss.html">Discussion</a></li>
<li><a class="dropdown-item" href="reference.html">Glossary</a></li>
          </ul>
</li>
      </ul>
</div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a class="btn btn-primary" href="aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div>
<!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Snakemake for Bioinformatics
</div>

<aside class="col-md-12 lesson-progress"><div style="width: %" class="percentage">
    %
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: %" aria-valuenow="" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/aio.html">Instructor View</a>
                      </div>
                    </div>
                  </div>
<!--/div.accordion-item-->
                </div>
<!--/div.accordion-flush-->
              </div>
<!--div.sidenav-view-selector -->
            </div>
<!--/div.col -->
      
            <hr>
</div>
<!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-introduction.html">1. Running commands with Snakemake</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-placeholders.html">2. Placeholders and wildcards</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-chaining_rules.html">3. Chaining rules</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-the_dag.html">4. How Snakemake plans its jobs</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-expansion.html">5. Processing lists of inputs</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="06-awkward_programs.html">6. Handling awkward programs</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="07-finished-workflow.html">7. Finishing the basic workflow</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="08-configuring.html">8. Configuring workflows</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush10">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading10">
        <a href="09-performance.html">9. Optimising workflow performance</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush11">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading11">
        <a href="10-conda_integration.html">10. Conda integration</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush12">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading12">
        <a href="11-assembly_challenge.html">11. Constructing a whole new workflow</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush13">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading13">
        <a href="12-cleaning_up.html">12. Cleaning up</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush14">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading14">
        <a href="13-quoting.html">13. Robust quoting in Snakefiles</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width">
<div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul>
<li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="discuss.html">Discussion</a></li>
<li><a href="reference.html">Glossary</a></li>
                    </ul>
</div>
                </div>
              </div>
            </div>
            <hr class="half-width resources">
<a href="aio.html">See all in one page</a>
            

            <hr class="d-none d-sm-block d-md-none">
<div class="d-grid gap-1">
            
            </div>
          </div>
<!-- /div.accordion -->
        </div>
<!-- /div.sidebar-inner -->
      </nav>
</div>
<!-- /div.sidebar -->
  </div>
<!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-extra.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <main id="main-content" class="main-content"><div class="container lesson-content">
        
        
<section id="aio-01-introduction"><p>Content from <a href="01-introduction.html">Running commands with Snakemake</a></p>
<hr>
<p>Last updated on 2024-02-22 |
        
        <a href="https://github.com/carpentries-incubator/snakemake-novice-bioinformatics/edit/main/episodes/01-introduction.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do I run a simple command with Snakemake?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Create a Snakemake recipe (a Snakefile)</li>
<li>Use Snakemake to count the lines in a FASTQ file</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="looking-at-the-sample-data"><h2 class="section-heading">Looking at the sample data<a class="anchor" aria-label="anchor" href="#looking-at-the-sample-data"></a>
</h2>
<hr class="half-width">
<p>You should have the sample data files unpacked already (if not, refer
back to the lesson setup). Under <code>yeast/reads</code> you’ll see a
number of FASTQ files (extension <code>.fq</code>). These represent
short paired RNA-seq reads from an experiment on yeast cultured under
three experimental conditions. For now we’ll just look at one single
file, <code>ref1_1.fq</code>.</p>
<p>In the terminal:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> cd yeast</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> ls reads</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> head <span class="at">-n8</span> reads/ref1_1.fq</span></code></pre>
</div>
<p>Files in FASTQ format have 4 lines for each sequence.</p>
<ul>
<li>Header line, beginning with <code>@</code>
</li>
<li>Sequence</li>
<li><code>+</code></li>
<li>Encoded quality per base</li>
</ul>
<p>Let’s count the number of lines in the file with <code>wc -l</code>.
Again, in the shell:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> wc <span class="at">-l</span> reads/ref1_1.fq</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="ex">58708</span></span></code></pre>
</div>
<p>We can redirect this result to a file like so:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> wc <span class="at">-l</span> reads/ref1_1.fq <span class="op">&gt;</span> ref1_1.fq.count</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> head <span class="at">-v</span> <span class="pp">*</span>.count</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="ex">==</span><span class="op">&gt;</span> ref1_1.fq.count <span class="op">&lt;</span>==</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="ex">58708</span> reads/ref1_1.fq</span></code></pre>
</div>
<div id="the-sample-dataset" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="the-sample-dataset" class="callout-inner">
<h3 class="callout-title">The sample dataset<a class="anchor" aria-label="anchor" href="#the-sample-dataset"></a>
</h3>
<div class="callout-content">
<p>The sample dataset represents a transcriptomics experiment in
brewer’s yeast (Saccharomyces cerevisiae) under three conditions:</p>
<ul>
<li>
<strong>etoh60</strong> - Treated with 60% ethanol</li>
<li>
<strong>temp33</strong> - Treated at 33 degrees celsius</li>
<li>
<strong>ref</strong> - Untreated</li>
</ul>
<p>For each condition there are 3 repeats, making 9 total samples. For
each, total RNA (or rather, cDNA) was sequenced on an Illumina HiSeq
instrument. For each repeat there is a pair of files as the sequencing
is double-ended, so for example <code>reads/etoh60_3_2.fq</code>
contains the second of the read pairs from the third ethanol-treated
sample.</p>
<p>Don’t worry about the biological meaning of this set-up. In the
course, we’re only going to get as far as assessing the quality of the
data and a very preliminary analysis. The data has been subsampled to a
fraction of the original size to make filtering and alignment operations
fast, so any deeper analysis is not going to yield meaningful
results.</p>
<p>As well as the reads, we have a transcriptome for the yeast. This
comes in a single FASTA file (as opposed to the cDNA reads which are in
FASTQ format) which has been GZip compressed.</p>
</div>
</div>
</div>
</section><section id="making-a-snakefile"><h2 class="section-heading">Making a Snakefile<a class="anchor" aria-label="anchor" href="#making-a-snakefile"></a>
</h2>
<hr class="half-width">
<p>Within the <code>yeast</code> directory, edit a new text file named
<code>Snakefile</code>.</p>
<p>Contents of <code>Snakefile</code>:</p>
<pre class="source"><code>rule countlines:
    output: "ref1_1.fq.count"
    input:  "reads/ref1_1.fq"
    shell:
        "wc -l reads/ref1_1.fq &gt; ref1_1.fq.count"</code></pre>
<div id="key-points-about-this-file" class="callout checklist">
<div class="callout-square">
<i class="callout-icon" data-feather="check-square"></i>
</div>
<div id="key-points-about-this-file" class="callout-inner">
<h3 class="callout-title">Key points about this file<a class="anchor" aria-label="anchor" href="#key-points-about-this-file"></a>
</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>The file is named <code>Snakefile</code> - with a capital
<code>S</code> and no file extension.</li>
<li>Some lines are indented. Indents must be with space characters, not
tabs. See the <a href="index.html#setup">setup section</a> for how to make
your text editor do this.</li>
<li>The rule definition starts with the keyword <code>rule</code>
followed by the rule name, then a colon.</li>
<li>We named the rule <code>countreads</code>. You may use letters,
numbers or underscores, but the rule name must begin with a letter and
may not be a keyword.</li>
<li>The keywords <code>input</code>, <code>output</code>,
<code>shell</code> are all followed by a colon.</li>
<li>The file names and the shell command are all in
<code>"quotes"</code>.</li>
<li>The output filename is given before the input filename. In fact,
Snakemake doesn’t care what order they appear in but we give the output
first throughout this course. We’ll see why soon.</li>
</ol>
</div>
</div>
</div>
<p>Back in the shell we’ll run our new rule. At this point, if there
were any missing quotes, bad indents, etc. we may see an error.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-j1</span> <span class="at">-F</span> <span class="at">-p</span> ref1_1.fq.count</span></code></pre>
</div>
<div id="running-snakemake" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="running-snakemake" class="callout-inner">
<h3 class="callout-title">Running Snakemake<a class="anchor" aria-label="anchor" href="#running-snakemake"></a>
</h3>
<div class="callout-content">
<p>Run <code>snakemake --help | less</code> to see the help for all
available options. What does the <code>-p</code> option in the
<code>snakemake</code> command above do?</p>
<ol style="list-style-type: decimal">
<li>Protects existing output files</li>
<li>Prints the shell commands that are being run to the terminal</li>
<li>Tells Snakemake to only run one process at a time</li>
<li>Prompts the user for the correct input file</li>
</ol>
<p><em>Hint: you can search in the text by pressing <code>/</code>, and
quit back to the shell with <code>q</code></em></p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">Show me the solution</h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<ol start="2" style="list-style-type: decimal">
<li>Prints the shell commands that are being run to the terminal</li>
</ol>
<p>This is such a useful thing we don’t know why it isn’t the default!
The <code>-j1</code> option is what tells Snakemake to only run one
process at a time, and we’ll stick with this for now as it makes things
simpler. The <code>-F</code> option tells Snakemake to always overwrite
output files, and we’ll learn about protected outputs much later in the
course. Answer 4 is a total red-herring, as Snakemake never prompts
interactively for user input.</p>
</div>
</div>
</div>
</div>
</section><section id="counting-sequences-in-a-fastq-file"><h2 class="section-heading">Counting sequences in a FASTQ file<a class="anchor" aria-label="anchor" href="#counting-sequences-in-a-fastq-file"></a>
</h2>
<hr class="half-width">
<p>FASTQ files contain 4 lines per sequence, as noted above. We can get
BASH to divide the output of <code>wc</code> by 4 to get the number of
sequences:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>$ echo $(( $(wc -l &lt;reads/ref1_1.fq) / 4 ))
14677</code></pre>
</div>
<p>Note that the input filename is now preceeded by <code>&lt;</code>.
This is a little trick to get <code>wc</code> to print only the number
of lines, and not the filename. The <code>$( ... )</code> syntax
captures this output value and the <code>$(( ... ))</code> syntax
encloses an arithmetic expression, which needs to be printed with
<code>echo</code>. Don’t worry if this is unfamiliar - you just need to
know that this is a shell command you can copy and use.</p>
<div id="counting-sequences-in-snakemake" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="counting-sequences-in-snakemake" class="callout-inner">
<h3 class="callout-title">Counting sequences in Snakemake<a class="anchor" aria-label="anchor" href="#counting-sequences-in-snakemake"></a>
</h3>
<div class="callout-content">
<p>Modify the Snakefile to count the number of
<strong>sequences</strong> in <code>reads/ref1_1.fq</code>, rather than
the number of <strong>lines</strong>.</p>
<ul>
<li>Rename the rule to “countreads”</li>
<li>Keep the output file name the same</li>
<li>Remember that the result needs to go into the output file, not just
be printed on the screen.</li>
</ul>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">Solution 1</h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<pre class="source"><code>rule countreads:
    output: "ref1_1.fq.count"
    input:  "reads/ref1_1.fq"
    shell:
        "echo $(( $(wc -l &lt;reads/ref1_1.fq) / 4 )) &gt; ref1_1.fq.count"</code></pre>
</div>
</div>
</div>
</div>
<div id="counting-sequences-in-snakemake" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="counting-sequences-in-snakemake" class="callout-inner">
<h3 class="callout-title">Counting sequences in Snakemake<em>(continued)</em><a class="anchor" aria-label="anchor" href="#counting-sequences-in-snakemake"></a>
</h3>
<div class="callout-content">
<p>Add a second rule to count the sequences in
<code>reads/etoh60_1_1.fq</code>. Add this to the same Snakefile you
already made, under the “countreads” rule, and run your rules in the
terminal. When running the <code>snakemake</code> command you’ll need to
tell Snakemake to make both the output files.</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">Solution 2</h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<p>You can choose whatever name you like for this second rule, but it
can’t be “countreads” as rule names need to be unique within a
Snakefile. So in this example answer we use “countreads2”.</p>
<pre class="source"><code>rule countreads2:
  output: "etoh60_1_1.fq.count"
  input:  "reads/etoh60_1_1.fq"
  shell:
    "echo $(( $(wc -l &lt;reads/etoh60_1_1.fq) / 4 )) &gt; etoh60_1_1.fq.count"</code></pre>
<p>Then in the shell…</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-j1</span> <span class="at">-F</span> <span class="at">-p</span> ref1_1.fq.count etoh60_1_1.fq.count</span></code></pre>
</div>
<p>If you think writing a separate rule for each output file is silly,
you are correct. We’ll address this next!</p>
</div>
</div>
</div>
</div>
<p><em>For reference, <a href="files/ep01.Snakefile">this is the full
Snakefile</a> we should have by the end of this episode.</em></p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Before running Snakemake you need to write a Snakefile</li>
<li>A Snakefile is a text file which defines a list of rules</li>
<li>Rules have inputs, outputs, and shell commands to be run</li>
<li>You tell Snakemake what file to make and it will run the shell
command defined in the appropriate rule</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-02-placeholders"><p>Content from <a href="02-placeholders.html">Placeholders and wildcards</a></p>
<hr>
<p>Last updated on 2024-02-23 |
        
        <a href="https://github.com/carpentries-incubator/snakemake-novice-bioinformatics/edit/main/episodes/02-placeholders.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do I make a generic rule?</li>
<li>How does Snakemake decide what rule to run?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Use Snakemake to count the sequences in any file</li>
<li>Understand the basic steps Snakemake goes through when running a
workflow</li>
<li>See how Snakemake deals with some errors</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<p><em>For reference, <a href="files/ep01.Snakefile">this is the
Snakefile</a> you should have to start the episode.</em></p>
<section id="wildcards-and-placeholders"><h2 class="section-heading">Wildcards and placeholders<a class="anchor" aria-label="anchor" href="#wildcards-and-placeholders"></a>
</h2>
<hr class="half-width">
<p>In the previous episode you wrote two rules to count the sequences in
two files. These work, but they are not a very efficient use of
Snakemake. We have eighteen input files to process and we don’t want to
write eighteen near-identical rules!</p>
<p>To make a more general-purpose rule we need
<strong>placeholders</strong> and <strong>wildcards</strong>. Here is a
new rule that will count the sequences in <strong>any</strong> of the
<code>.fq</code> files.</p>
<pre class="source"><code># New generic read counter
rule countreads:
    output: "{myfile}.fq.count"
    input:  "reads/{myfile}.fq"
    shell:
        "echo $(( $(wc -l &lt;{input}) / 4 )) &gt; {output}"</code></pre>
<div id="comments-in-snakefiles" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="comments-in-snakefiles" class="callout-inner">
<h3 class="callout-title">Comments in Snakefiles<a class="anchor" aria-label="anchor" href="#comments-in-snakefiles"></a>
</h3>
<div class="callout-content">
<p>In the above code, the line beginning <code>#</code> is a comment
line. Hopefully you are already in the habit of adding comments to your
own scripts. Good comments make any script more readable, and this is
just as true with Snakefiles.</p>
</div>
</div>
</div>
<p>As a reminder, here’s the non-generic version from the last
episode:</p>
<pre class="source"><code># Original version
rule countreads:
    output: "ref1_1.fq.count"
    input:  "reads/ref1_1.fq"
    shell:
        "echo $(( $(wc -l &lt;reads/ref1_1.fq) / 4 )) &gt; ref1_1.fq.count"</code></pre>
<p>The new rule has replaced explicit file names with things in
<code>{curly brackets}</code>, specifically <code>{myfile}</code>,
<code>{input}</code> and <code>{output}</code>.</p>
<div class="section level3">
<h3 id="myfile-is-a-wildcard">
<code>{myfile}</code> is a <strong>wildcard</strong>
<a class="anchor" aria-label="anchor" href="#myfile-is-a-wildcard"></a>
</h3>
<p>Wildcards are used in the <code>input</code> and <code>output</code>
lines of the rule to represent parts of filenames. Much like the
<code>*</code> pattern in the shell, the wildcard can stand in for any
text in order to make up the desired filename. As with naming your
rules, you may choose any name you like for your wildcards, so here we
used <code>myfile</code>. If <code>myfile</code> is set to
<code>ref1_1</code> then the new generic rule will have the same inputs
and outputs as the original rule. Using the same wildcards in the input
and output is what tells Snakemake how to match input files to output
files.</p>
<p>If two rules use a wildcard with the same name then Snakemake will
treat them as different entities</p>
<ul>
<li>rules in Snakemake are self-contained in this way.</li>
</ul>
</div>
<div class="section level3">
<h3 id="input-and-output-are-placeholders">
<code>{input}</code> and <code>{output}</code> are
<strong>placeholders</strong>
<a class="anchor" aria-label="anchor" href="#input-and-output-are-placeholders"></a>
</h3>
<p>Placeholders are used in the <code>shell</code> section of a rule,
and Snakemake will replace them with appropriate values -
<code>{input}</code> with the full name of the input file, and
<code>{output}</code> with the full name of the output file – before
running the command.</p>
<p>If we had wanted to include the value of the <code>myfile</code>
wildcard directly in the <code>shell</code> command we could have used
the placeholder <code>{wildcards.myfile}</code> but in most cases, as
here, we just need the <code>{input}</code> and <code>{output}</code>
placeholders.</p>
<div id="running-the-general-purpose-rule" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="running-the-general-purpose-rule" class="callout-inner">
<h3 class="callout-title">Running the general-purpose rule<a class="anchor" aria-label="anchor" href="#running-the-general-purpose-rule"></a>
</h3>
<div class="callout-content">
<p>Modify your Snakefile to incorporate the changes described above,
using the wildcard and input/output placeholders. You should resist the
urge to copy-and-paste from this workbook, but rather edit the file by
hand, as this will stick better in your memory.</p>
<p>You should delete the now-redundant second rule, so your Snakefile
should contain just one rule named <em>countreads</em>.</p>
<p>Using this new rule, determine: how many reads are in the
<code>temp33_1_1.fq</code> file?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">Show me the solution</h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>After editing the file, run the commands:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-j1</span> <span class="at">-F</span> <span class="at">-p</span> temp33_1_1.fq.count</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> cat temp33_1_1.fq.count</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="ex">20593</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="choosing-the-right-wildcards" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="choosing-the-right-wildcards" class="callout-inner">
<h3 class="callout-title">Choosing the right wildcards<a class="anchor" aria-label="anchor" href="#choosing-the-right-wildcards"></a>
</h3>
<div class="callout-content">
<p>Our rule puts the sequence counts into output files named like
<code>ref1_1.fq.count</code>. How would you have to change the
“countreads” rule definition if you wanted:</p>
<ol style="list-style-type: decimal">
<li><p>the output file for <code>reads/ref1_1.fq</code> to be
<code>counts/ref1_1.txt</code>?</p></li>
<li><p>the output file for <code>reads/ref1_1.fq</code> to be
<code>ref1_counts/fq.1.count</code> (for <code>reads/ref1_2.fq</code> to
be <code>ref1_counts/fq.2.count</code>, etc.)?</p></li>
<li><p>the output file for <code>reads/ref1_1.fq</code> to be
<code>countreads_1.txt</code>?</p></li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">Show me the solution</h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<p>In all cases, there is no need to change the <code>shell</code> part
of the rule at all.</p>
<ol style="list-style-type: decimal"><li>
</li></ol>
<pre class="source"><code><span><span class="va">output</span><span class="op">:</span> <span class="st">"counts/{myfile}.txt"</span></span>
<span><span class="va">input</span><span class="op">:</span>  <span class="st">"reads/{myfile}.fq"</span></span></code></pre>
<p>This can be done just by changing the <code>output:</code> line. You
may also have considered the need to <code>mkdir counts</code> but in
fact this is not necessary as Snakemake will create the output directory
path for you before it runs the rule.</p>
<ol start="2" style="list-style-type: decimal"><li>
</li></ol>
<pre class="source"><code><span><span class="va">output</span><span class="op">:</span> <span class="st">"{sample}_counts/fq.{readnum}.count"</span></span>
<span><span class="va">input</span><span class="op">:</span>  <span class="st">"reads/{sample}_{readnum}.fq"</span></span></code></pre>
<p>In this case, it was necessary to introduce a second wildcard,
because the elements in the output file name are split up. The names
chosen here are <code>{sample}</code> and <code>{readnum}</code> but you
could choose any names as long as they match between the
<code>input</code> and <code>output</code> parts. Once again, the output
directory will be created for us by Snakemake, so the <code>shell</code>
command does not need to change.</p>
<ol start="3" style="list-style-type: decimal"><li>
</li></ol>
<p>This one <strong>isn’t possible</strong>, because Snakemake cannot
determine which input file you want to count by matching wildcards on
the file name “countreads_1.txt”. You could try a rule like this:</p>
<pre class="source"><code><span><span class="va">output</span><span class="op">:</span> <span class="st">"countreads_{readnum}.count"</span></span>
<span><span class="va">input</span><span class="op">:</span>  <span class="st">"reads/ref1_{readnum}.fq"</span></span></code></pre>
<p>…but it only works because “ref1” is hard-coded into the
<code>input</code> line, and the rule will only work on this specific
sample, not the other eight in our sample dataset. In general, input and
output filenames need to be carefully chosen so that Snakemake can match
everything up and determine the right input from the output
filename.</p>
</div>
</div>
</div>
</div>
</div>
</section><section id="snakemake-order-of-operations"><h2 class="section-heading">Snakemake order of operations<a class="anchor" aria-label="anchor" href="#snakemake-order-of-operations"></a>
</h2>
<hr class="half-width">
<p>We’re only just getting started with some simple rules, but it’s
worth thinking about exactly what Snakemake is doing when you run it.
There are three distinct phases:</p>
<ol style="list-style-type: decimal">
<li>Prepares to run:
<ol style="list-style-type: decimal">
<li>Reads in all the rule definitions from the Snakefile</li>
</ol>
</li>
<li>Plans what to do:
<ol style="list-style-type: decimal">
<li>Sees what file(s) you are asking it to make</li>
<li>Looks for a matching rule by looking at the <code>output</code>s of
all the rules it knows</li>
<li>Fills in the wildcards to work out the <code>input</code> for this
rule</li>
<li>Checks that this input file is actually available</li>
</ol>
</li>
<li>Runs the steps:
<ol style="list-style-type: decimal">
<li>Creates the directory for the output file, if needed</li>
<li>Removes the old output file if it is already there</li>
<li>Only then, runs the shell command with the placeholders
replaced</li>
<li>Checks that the command ran without errors <em>and</em> made the new
output file as expected</li>
</ol>
</li>
</ol>
<p>For example, if we now ask Snakemake to generate a file named
<code>wibble_1.fq.count</code>:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>$ snakemake -j1 -F -p wibble_1.fq.count
Building DAG of jobs...
MissingInputException in line 1 of /home/zenmaster/data/yeast/Snakefile:
Missing input files for rule countreads:
reads/wibble_1.fq</code></pre>
</div>
<p>Snakemake sees that a file with a name like this could be produced by
the <em>countreads</em> rule. However, when it performs the wildcard
substitution it sees that the input file would need to be named
<code>reads/wibble_1.fq</code>, and there is no such file available.
Therefore Snakemake stops and gives an error before any shell commands
are run.</p>
<div id="dry-run--n-mode" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="dry-run--n-mode" class="callout-inner">
<h3 class="callout-title">Dry-run (-n) mode<a class="anchor" aria-label="anchor" href="#dry-run--n-mode"></a>
</h3>
<div class="callout-content">
<p>It’s often useful to run just the first two phases, so that Snakemake
will plan out the jobs to run, and print them to the screen, but never
actually run them. This is done with the <code>-n</code> flag, eg:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-n</span> <span class="at">-F</span> <span class="at">-p</span> temp33_1_1.fq.count</span></code></pre>
</div>
<p>We’ll make use of this later in the course.</p>
</div>
</div>
</div>
<p>The amount of checking may seem pedantic right now, but as the
workflow gains more steps this will become very useful to us indeed.</p>
<div id="adding-a-second-rule" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="adding-a-second-rule" class="callout-inner">
<h3 class="callout-title">Adding a second rule<a class="anchor" aria-label="anchor" href="#adding-a-second-rule"></a>
</h3>
<div class="callout-content">
<p>Here is a command that will trim and filter low quality reads from a
FASTQ file.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> fastq_quality_trimmer <span class="at">-t</span> 20 <span class="at">-l</span> 100 <span class="at">-o</span> output.fq <span class="op">&lt;</span>input.fq</span></code></pre>
</div>
<p>Add a second rule to your Snakefile to run this trimmer. You should
make it so that valid outputs are files with the same name as the input,
but in a subdirectory named ‘trimmed’, for example:</p>
<ul>
<li>trimmed/ref1_1.fq</li>
<li>trimmed/temp33_1_1.fq</li>
<li><em>etc.</em></li>
</ul>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">Show me the solution</h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<pre class="source"><code># Trim any FASTQ reads for base quality
rule trimreads:
    output: "trimmed/{myfile}.fq"
    input:  "reads/{myfile}.fq"
    shell:
        "fastq_quality_trimmer -t 20 -l 100 -o {output} &lt;{input}"</code></pre>
<p>Bonus points if you added any comments to the code!</p>
<p>And of course you can run your new rule as before, to make one or
more files at once. For example:</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-j1</span> <span class="at">-F</span> <span class="at">-p</span> trimmed/ref1_1.fq trimmed/ref1_2.fq</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="about-fastq_quality_trimmer" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="about-fastq_quality_trimmer" class="callout-inner">
<h3 class="callout-title">About fastq_quality_trimmer<a class="anchor" aria-label="anchor" href="#about-fastq_quality_trimmer"></a>
</h3>
<div class="callout-content">
<p><code>fastq_quality_trimmer</code> is part of the <a href="https://hannonlab.cshl.edu/fastx_toolkit/" class="external-link">FastX toolkit</a> and
performs basic trimming on single FASTQ files. The options
<code>-t 20 -l 100</code> happen to be reasonable quality cutoffs for
this dataset. This program reads from standard input so we’re using
<code>&lt;</code> to specify the input file, and the <code>-o</code>
flag specifies the output name.</p>
</div>
</div>
</div>
<p><em>For reference, <a href="files/ep02.Snakefile">this is a
Snakefile</a> incorporating the changes made in this episode.</em></p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Snakemake rules are made generic with placeholders and
wildcards</li>
<li>Snakemake chooses the appropriate rule by replacing wildcards such
the the output matches the target</li>
<li>Placeholders in the shell part of the rule are replaced with values
based on the chosen wildcards</li>
<li>Snakemake checks for various error conditions and will stop if it
sees a problem</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-03-chaining_rules"><p>Content from <a href="03-chaining_rules.html">Chaining rules</a></p>
<hr>
<p>Last updated on 2024-03-01 |
        
        <a href="https://github.com/carpentries-incubator/snakemake-novice-bioinformatics/edit/main/episodes/03-chaining_rules.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do I combine rules into a workflow?</li>
<li>How do I make a rule with multiple inputs and outputs?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Use Snakemake to filter and then count the lines in a FASTQ
file</li>
<li>Add an RNA quantification step in the data analysis</li>
<li>See how Snakemake deals with missing outputs</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<p><em>For reference, <a href="files/ep02.Snakefile">this is the
Snakefile</a> you should have to start the episode.</em></p>
<section id="a-pipeline-of-multiple-rules"><h2 class="section-heading">A pipeline of multiple rules<a class="anchor" aria-label="anchor" href="#a-pipeline-of-multiple-rules"></a>
</h2>
<hr class="half-width">
<p>We now have a “trimreads” rule and a “countreads” rule. Following the
previous chapter, the contents of the Snakefile should be:</p>
<pre class="source"><code># New generic read counter
rule countreads:
    output: "{myfile}.fq.count"
    input:  "reads/{myfile}.fq"
    shell:
        "echo $(( $(wc -l &lt;{input}) / 4 )) &gt; {output}"

# Trim any FASTQ reads for base quality
rule trimreads:
    output: "trimmed/{myfile}.fq"
    input:  "reads/{myfile}.fq"
    shell:
        "fastq_quality_trimmer -t 20 -l 100 -o {output} &lt;{input}"</code></pre>
<p>The problem is there is no good way to use these rules together, that
is, to trim an input file and then count the reads in the trimmed file.
The <em>countreads</em> rule only takes input reads from the
<em>reads</em> directory, whereas the <em>trimreads</em> rule puts all
results into the <em>trimmed</em> directory.</p>
<p>Chaining rules in Snakemake is a matter of choosing filename patterns
that connect the rules. There’s something of an art to it - most times
there are several options that will work. Consider the following
alternative version of the <em>countreads</em> rule:</p>
<pre class="source"><code># New even-more-generic read counter
rule countreads:
    output: "{indir}.{myfile}.fq.count"
    input:  "{indir}/{myfile}.fq"
    shell:
        "echo $(( $(wc -l &lt;{input}) / 4 )) &gt; {output}"</code></pre>
<p>Now, the rule no longer requires the input files to be in the “reads”
directory. The directory name has been replaced by the
<code>{indir}</code> wildcard. We can request Snakemake to create a file
following this output pattern:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-j1</span> <span class="at">-F</span> <span class="at">-p</span> trimmed.ref1_1.fq.count</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> cat trimmed.ref1_1.fq.count</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="ex">14278</span></span></code></pre>
</div>
<p>Look at the logging messages that Snakemake prints in the terminal.
What has happened here?</p>
<ol style="list-style-type: decimal">
<li>Snakemake looks for a rule to make
<code>trimmed.ref1_1.fq.count</code>
</li>
<li>It determines that “countreads” can make this if
<code>indir=trimmed</code> and <code>myfile=ref1_1</code>
</li>
<li>It sees that the input needed is therefore
<code>trimmed/ref1_1.fq</code> <br><br>
</li>
<li>Snakemake looks for a rule to make
<code>trimmed/ref1_1.fq</code>
</li>
<li>It determines that “trimreads” can make this if
<code>myfile=ref1_1</code>
</li>
<li>It sees that the input needed is therefore
<code>reads/ref1_1.fq</code> <br><br>
</li>
<li>Now Snakemake has reached an available input file, it runs both
steps to get the final output</li>
</ol>
<p><strong>Here’s a visual representation of this process:</strong></p>
<figure><img src="fig/chaining_rules.png" alt="A visual representation of the above process showing the rule definitions, with arrows added to indicate the order wildcards and placeholders are substituted. Blue arrows start from the final target at the top, which is the file trimmed.ref1_1.fq.count, then point down from components of the filename to wildcards in the output of the countreads rule. Arrows from the input of this rule go down to the output of the trimreads rule. Orange arrows then track back up through the shell parts of both rules, where the placeholders are, and finally back to the target output filename at the top." class="figure mx-auto d-block"></figure><p>This, in a nutshell, is how we build workflows in Snakemake.</p>
<ol style="list-style-type: decimal">
<li>Define rules for all the processing steps</li>
<li>Choose <code>input</code> and <code>output</code> naming patterns
that allow Snakemake to link the rules</li>
<li>Tell Snakemake to generate the final output files</li>
</ol>
<p>If you are used to writing regular scripts this takes a little
getting used to. Rather than listing steps in order of execution, you
are always <strong>working backwards</strong> from the final desired
result. The order of operations is determined by applying the pattern
matching rules to the filenames, not by the order of the rules in the
Snakefile.</p>
<div id="outputs-first" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="outputs-first" class="callout-inner">
<h3 class="callout-title">Outputs first?<a class="anchor" aria-label="anchor" href="#outputs-first"></a>
</h3>
<div class="callout-content">
<p>The Snakemake approach of working backwards from the desired output
to determine the workflow is why we’re putting the <code>output</code>
lines first in all our rules - to remind us that these are what
Snakemake looks at first!</p>
<p>Many users of Snakemake, and indeed the official documentation,
prefer to have the <code>input</code> first, so in practise you should
use whatever order makes sense to you.</p>
</div>
</div>
</div>
<div id="thinking-about-your-own-workflows" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="thinking-about-your-own-workflows" class="callout-inner">
<h3 class="callout-title">Thinking about your own workflows<a class="anchor" aria-label="anchor" href="#thinking-about-your-own-workflows"></a>
</h3>
<div class="callout-content">
<p>Think about any data processing task you have done yourself, and
write down three or four steps from that workflow.</p>
<p>What were the inputs to, and outputs from, each step?</p>
<p>How did the steps connect up, in terms of data going from one to the
next? You may want to sketch this out and use arrows to indicate the
linkages between the steps.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">A sample answer based on brewing a mug of tea.</h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p><em>TODO - draw this out too.</em></p>
<pre><code>"Boil Water" : input=cold water, output=hot water

"Brew Tea" : input=[hot water, tea bag, mug], output=tea infusion

"Add Milk And Sugar" : input=[tea infusion, milk, sugar], output=cuppa</code></pre>
</div>
</div>
</div>
</div>
</section><section id="adding-an-alignment-step-to-the-pipeline"><h2 class="section-heading">Adding an alignment step to the pipeline<a class="anchor" aria-label="anchor" href="#adding-an-alignment-step-to-the-pipeline"></a>
</h2>
<hr class="half-width">
<p>Let’s add another rule to our Snakefile. The reads we have are from a
yeast RNA-seq experiment so we might reasonably want to quantify
transcript abundance using the <strong>kallisto</strong> aligner. The
command to do so looks like this:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> kallisto quant <span class="at">-i</span> index_file <span class="at">-o</span> output_dir in_1.fastq in_2.fastq</span></code></pre>
</div>
<p>This command has three input files:</p>
<ol style="list-style-type: decimal">
<li>The transcriptome index</li>
<li>The first of the paired FASTQ files</li>
<li>The second of the paired FASTQ files</li>
</ol>
<p>And it produces a directory of output files. According to <a href="https://pachterlab.github.io/kallisto/manual#quant" class="external-link">the Kallisto
manual</a> this directory will have three output files in it:</p>
<ol style="list-style-type: decimal">
<li>abundance.h5</li>
<li>abundance.tsv</li>
<li>run_info.json</li>
</ol>
<p>We’ll not worry about what the contents of these files mean just now,
or how Kallisto generates them. We just know that we want to run the
<code>kallisto quant</code> command and have it make the output files,
and the output files are going to be useful once we add later steps in
the analysis.</p>
<p>Making a rule with multiple inputs and outputs like this works much
like the previous rules.</p>
<pre class="source"><code>rule kallisto_quant:
    output:
        h5   = "kallisto.{sample}/abundance.h5",
        tsv  = "kallisto.{sample}/abundance.tsv",
        json = "kallisto.{sample}/run_info.json",
    input:
        index = "Saccharomyces_cerevisiae.R64-1-1.kallisto_index",
        fq1   = "trimmed/{sample}_1.fq",
        fq2   = "trimmed/{sample}_2.fq",
    shell:
        "kallisto quant -i {input.index} -o kallisto.{wildcards.sample} {input.fq1} {input.fq2}"</code></pre>
<p>There are many things to note here:</p>
<ol style="list-style-type: decimal">
<li>The individual input and output files are given names using the
<code>=</code> syntax.</li>
<li>Each of these lines must end with a <code>,</code> (optional for the
last one).</li>
<li>In the <code>shell</code> part, the input placeholders are now like
<code>{input.name}</code>.</li>
<li>We’ve chosen to only quantify the <em>trimmed</em> version of the
reads.</li>
<li>We’ve used the wildcard name <code>{sample}</code> rather than
<code>{myfile}</code> because this will match only the sample name, eg
<code>ref1</code>, not <code>ref1_1</code>. Snakemake doesn’t care what
name we use, but carefully chosen names make for more readable
rules.</li>
<li>Because <code>kallisto quant</code> only takes the output directory
name, we’ve used the placeholder <code>{wildcards.sample}</code> rather
than <code>{output}</code> which would give the full file names.</li>
<li>We don’t actually have the <code>{input.index}</code> file yet. This
will need to be created using the <code>kallisto index</code>
command.</li>
<li>If the number of input or output files had been variable, we’d need
a slightly different approach. We’ll come on to this in a later
episode.</li>
</ol>
<p>Even though the rule is not going to work without the index, we can
still run it to check that Snakemake is happy with the rule
definition.</p>
<div id="running-kallisto-on-all-replicates" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="running-kallisto-on-all-replicates" class="callout-inner">
<h3 class="callout-title">Running Kallisto on all replicates<a class="anchor" aria-label="anchor" href="#running-kallisto-on-all-replicates"></a>
</h3>
<div class="callout-content">
<p>If you know about the Kallisto software, you may be thinking about
running Kallisto on all replicates of the condition at once. We’ll look
at this later in the course, but for now assume that Kallisto is run
once per sample, ie. once for each pair of FASTQ files.</p>
</div>
</div>
</div>
<div id="running-the-kallisto_quant-rule" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="running-the-kallisto_quant-rule" class="callout-inner">
<h3 class="callout-title">Running the kallisto_quant rule<a class="anchor" aria-label="anchor" href="#running-the-kallisto_quant-rule"></a>
</h3>
<div class="callout-content">
<p>Given that the <em>index</em> input is missing, what would you expect
Snakemake to do if the new rule was run now?</p>
<p>Try it by telling Snakemake to run the new rule on the files
<code>ref1_1.fq</code> and <code>ref1_2.fq</code>. Since the rule
defines multiple outputs, asking for any one of the output files will be
enough.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">Show me the solution</h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-j1</span> <span class="at">-F</span> <span class="at">-p</span> kallisto.ref1/abundance.h5</span></code></pre>
</div>
<p>Resulting error is “Missing input files”. Note that Snakemake does
not try to run kallisto at all. It stops while still making a work plan,
because it sees that a required input file is missing.</p>
</div>
</div>
</div>
</div>
<div id="building-the-index" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="building-the-index" class="callout-inner">
<h3 class="callout-title">Building the index<a class="anchor" aria-label="anchor" href="#building-the-index"></a>
</h3>
<div class="callout-content">
<p>Instruct Snakemake how to build the genome index as part of the
pipeline by adding another rule. The command we need to run is:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> kallisto index <span class="at">-i</span> index_file_to_make fasta_file_to_index</span></code></pre>
</div>
<p>The file to be indexed is
<code>transcriptome/Saccharomyces_cerevisiae.R64-1-1.cdna.all.fa.gz</code>.
As there is only one input to the rule you don’t have to give it a name,
but you may do so if you like.</p>
<p>Make it so that the output printed by the program is captured to a
file, and therefore your rule will have two separate outputs: the
<em>index file</em> and the <em>log file</em>. Note that the program
prints messages on <em>stderr</em>, so you will need to use
<code>&gt;&amp;</code> rather than <code>&gt;</code> to capture the
output.</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">Show me the solution</h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<p>The rule you write could look something like this, but there are many
variations that will work just as well. Since there is only one
transcriptome in the project, you may feel that use of the
<code>{strain}</code> wildcard is overkill, but who’s to say we might
not want to use another in future?</p>
<pre class="source"><code>rule kallisto_index:
    output:
        idx = "{strain}.kallisto_index",
        log = "{strain}.kallisto_log",
    input:
        fasta = "transcriptome/{strain}.cdna.all.fa.gz"
    shell:
        "kallisto index -i {output.idx} {input.fasta} &gt;&amp; {output.log}"</code></pre>
</div>
</div>
</div>
</div>
<div id="log-outputs-in-snakemake" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="log-outputs-in-snakemake" class="callout-inner">
<h3 class="callout-title">
<code>log</code>outputs in Snakemake<a class="anchor" aria-label="anchor" href="#log-outputs-in-snakemake"></a>
</h3>
<div class="callout-content">
<p>Snakemake has a dedicated rule field for outputs that are <a href="https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#log-files" class="external-link">log
files</a>, and these are mostly treated as regular outputs except that
log files are not removed if the job produces an error. This means you
can look at the log to help diagnose the error. In a real workflow this
can be very useful, but in terms of learning the fundementals of
Snakemake we’ll stick with regular <code>input</code> and
<code>output</code> fields here.</p>
</div>
</div>
</div>
</section><section id="dealing-with-a-missing-files-error"><h2 class="section-heading">Dealing with a <em>missing files</em> error<a class="anchor" aria-label="anchor" href="#dealing-with-a-missing-files-error"></a>
</h2>
<hr class="half-width">
<p>All being well, your new rules are now ready to run Kallisto.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-j1</span> <span class="at">-F</span> <span class="at">-p</span> kallisto.ref1/abundance.h5</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="ex">...lots</span> of output...</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="ex">4</span> of 4 steps <span class="er">(</span><span class="ex">100%</span><span class="kw">)</span> <span class="cf">done</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="ex">Complete</span> log: /home/zenmaster/data/yeast/.snakemake/log/2021-04-23T142649.632834.snakemake.log</span></code></pre>
</div>
<p>We’ll end the chapter by looking at a common problem that can arise
if you mistype a file name in a rule. Remember that we wrote the rule
based on the expected output filenames given in the Kallisto manual. In
an older version of this manual there was a typo where the file names
were incorrectly given as <code>abundances.h5</code> and
<code>abundances.tsv</code>, with the extra <code>s</code> on each.</p>
<p>It may seem silly to break the workflow when we just got it working,
but it will be instructive, so edit the Snakefile and change these names
to the incorrect versions.</p>
<pre class="source"><code>rule kallisto_quant:
    output:
        h5   = "kallisto.{sample}/abundances.h5",
        tsv  = "kallisto.{sample}/abundances.tsv",
        json = "kallisto.{sample}/run_info.json",
...</code></pre>
<p>To keep things tidy, this time we’ll manually remove the output
directory.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> rm <span class="at">-rvf</span> kallisto.ref1</span></code></pre>
</div>
<p>And re-run. Note that the output file name you’ll need to use on the
command line must match the edited Snakefile, or you will get a
<code>MissingRuleException</code>.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>$ snakemake -j1 -F -p kallisto.ref1/abundances.h5

...
kallisto quant -i Saccharomyces_cerevisiae.R64-1-1.kallisto_index -o kallisto.ref1 trimmed/ref1_2.fq trimmed/ref1_2.fq

[quant] fragment length distribution will be estimated from the data
...more Kallisto output...
[   em] the Expectation-Maximization algorithm ran for 265 rounds

Waiting at most 5 seconds for missing files.
MissingOutputException in line 24 of /home/zenmaster/data/yeast/Snakefile:
Job Missing files after 5 seconds:
kallisto.ref1/abundances.h5
kallisto.ref1/abundances.tsv
This might be due to filesystem latency. If that is the case, consider to increase the wait time with --latency-wait.
Job id: 0 completed successfully, but some output files are missing. 0
  File "/opt/python3.7/site-packages/snakemake/executors/__init__.py", line 583, in handle_job_success
  File "/opt/python3.7/site-packages/snakemake/executors/__init__.py", line 259, in handle_job_success
Removing output files of failed job kallisto_quant since they might be corrupted:
kallisto.ref1/run_info.json
Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
Complete log: /home/zenmaster/data/yeast/.snakemake/log/2021-04-23T142649.632834.snakemake.log</code></pre>
</div>
<p>There’s a lot to take in here. Some of the messages are very
informative. Some less so.</p>
<ol style="list-style-type: decimal">
<li>Snakemake did actually run kallisto, as evidenced by the output from
kallisto that we see on the screen.</li>
<li>There is no obvious error message being reported by kallisto.</li>
<li>Snakemake complains some expected output files are missing:
<code>kallisto.ref1/abundances.h5</code> and
<code>kallisto.ref1/abundances.tsv</code>.</li>
<li>The third expected output file
<code>kallisto.ref1/run_info.json</code> was found but has now been
removed by Snakemake.</li>
<li>Snakemake suggest this might be due to “filesystem latency”.</li>
</ol>
<p>This last point is a red herring. “Filesystem latency” is not an
issue here, and never will be since we are not using a network
filesystem. We know what the problem is, as we deliberately caused it,
but to diagnose an unexpected error like this we would investigate
further by looking at the <code>kallisto.ref1</code> subdirectory.</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> ls kallisto.ref1/</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="ex">abundance.h5</span>  abundance.tsv</span></code></pre>
</div>
<p>Remember that Snakemake itself does not create any output files. It
just runs the commands you put in the <code>shell</code> sections, then
checks to see if all the expected output files have appeared.</p>
<p>So if the file names created by kallisto are not exactly the same as
in the Snakefile you will get this error, and you will, in this case,
find that some output files are present but others
(<code>run_info.json</code>, which was named correctly) have been
cleaned up by Snakemake.</p>
<div id="errors-are-normal" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="errors-are-normal" class="callout-inner">
<h3 class="callout-title">Errors are normal<a class="anchor" aria-label="anchor" href="#errors-are-normal"></a>
</h3>
<div class="callout-content">
<p>Don’t be disheartened if you see errors like the one above when first
testing your new Snakemake pipelines. There is a lot that can go wrong
when writing a new workflow, and you’ll normally need several iterations
to get things just right. One advantage of the Snakemake approach
compared to regular scripts is that Snakemake fails fast when there is a
problem, rather than ploughing on and potentially running junk
calculations on partial or corrupted data. Another advantage is that
when a step fails we can safely resume from where we left off, as we’ll
see in the next episode.</p>
</div>
</div>
</div>
<p>Finally, edit the names in the Snakefile back to the correct version
and re-run to confirm that all is well. Assure yourself that that the
rules are still generic by processing the <em>temp33_1</em> sample
too:</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-j1</span> <span class="at">-F</span> <span class="at">-p</span> kallisto.ref1/abundance.h5  kallisto.temp33_1/abundance.h5</span></code></pre>
</div>
<p><em>For reference, <a href="files/ep03.Snakefile">this is a
Snakefile</a> incorporating the changes made in this episode.</em></p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Snakemake links rules by iteratively looking for rules that make
missing inputs</li>
<li>Rules may have multiple named inputs and/or outputs</li>
<li>If a shell command does not yield an expected output then Snakemake
will regard that as a failure</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-04-the_dag"><p>Content from <a href="04-the_dag.html">How Snakemake plans its jobs</a></p>
<hr>
<p>Last updated on 2024-03-11 |
        
        <a href="https://github.com/carpentries-incubator/snakemake-novice-bioinformatics/edit/main/episodes/04-the_dag.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do I visualise a Snakemake workflow?</li>
<li>How does Snakemake avoid unecessary work?</li>
<li>How do I control what steps will be run?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>View the DAG for our pipeline</li>
<li>Understand the logic Snakemake uses when running and re-running
jobs</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<p><em>For reference, <a href="files/ep03.Snakefile">this is the
Snakefile</a> you should have to start the episode.</em></p>
<section id="the-dag"><h2 class="section-heading">The DAG<a class="anchor" aria-label="anchor" href="#the-dag"></a>
</h2>
<hr class="half-width">
<p>You may have noticed that one of the messages Snakemake always prints
is:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Building DAG of jobs...</code></pre>
</div>
<p>A DAG is a <strong>Directed Acyclic Graph</strong> and it can be
pictured like so:</p>
<figure><img src="fig/dag_1.svg" alt="Diagram showing jobs as coloured boxes joined by arrows representing data flow. The box labelled as kallisto_index is in green at the top, with two blue boxes labelled trimreads and two yellow boxes labelled countreads. The blue trimreads boxes have arrows into the respective yellow countreads boxes. Finally there is a kallisto_quant job shown as a red box, with incoming arrows from both the trimreads box as well as the kallisto_index box." class="figure mx-auto d-block"></figure><p>The above DAG is based on our four existing rules, and shows all the
jobs Snakemake would run to trim, count and quantify the <em>ref1</em>
sample.</p>
<div id="note-that" class="callout checklist">
<div class="callout-square">
<i class="callout-icon" data-feather="check-square"></i>
</div>
<div id="note-that" class="callout-inner">
<h3 class="callout-title">Note that:<a class="anchor" aria-label="anchor" href="#note-that"></a>
</h3>
<div class="callout-content">
<ul>
<li>A rule can appear more than once, with different wildcards (a
<strong>rule</strong> plus <strong>wildcard values</strong> defines a
<strong>job</strong>)</li>
<li>The arrows show dependency ordering between jobs</li>
<li>Snakemake can run the jobs in any order that doesn’t break
dependency - for example <em>kallisto_quant</em> cannot run until both
<em>kallisto_index</em> and <em>trimreads</em> have completed, but it
may run before or after <em>countreads</em>
</li>
<li>This is a work list, <em>not a flowchart</em>, so there are no
if/else decisions or loops - Snakemake runs every job in the DAG exactly
once</li>
<li>The DAG depends both on the Snakefile <em>and</em> on the requested
target outputs, and the files already present</li>
<li>When building the DAG, Snakemake does not look at the <em>shell</em>
part of the rules at all - only when running the DAG will Snakemake
check that the shell commands are working and producing the expected
output files</li>
</ul>
</div>
</div>
</div>
<div id="how-many-jobs" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="how-many-jobs" class="callout-inner">
<h3 class="callout-title">How many jobs?<a class="anchor" aria-label="anchor" href="#how-many-jobs"></a>
</h3>
<div class="callout-content">
<p>If we asked Snakemake to run <em>kallisto_quant</em> on all three of
the reference samples (ref1, ref2, ref3), how many jobs would that be in
total?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">Show me the solution</h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>10 in total: 3 * kallisto_quant // 6 * trimreads // 1 *
kallisto_index // 0 * countreads</p>
</div>
</div>
</div>
</div>
</section><section id="snakemake-is-lazy-and-laziness-is-good"><h2 class="section-heading">Snakemake is lazy, and laziness is good<a class="anchor" aria-label="anchor" href="#snakemake-is-lazy-and-laziness-is-good"></a>
</h2>
<hr class="half-width">
<p>For the last few episodes, we’ve told you to run Snakemake like
this:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-j1</span> <span class="at">-F</span> <span class="at">-p</span> desired_output_file</span></code></pre>
</div>
<p>As a reminder, the <code>-j1</code> flag tells Snakemake to run one
job at a time, and <code>-p</code> is to print out the shell commands
before running them.</p>
<p>The <code>-F</code> flag turns on <code>forceall</code> mode, and in
normal usage you don’t want this.</p>
<p>At the end of the last chapter, we generated some kallisto results by
running:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-j1</span> <span class="at">-F</span> <span class="at">-p</span> kallisto.temp33_1/abundance.h5</span></code></pre>
</div>
<p>Now try without the <code>-F</code> option. Assuming that the output
files are already created, you’ll see this:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-j1</span> <span class="at">-p</span> kallisto.temp33_1/abundance.h5</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="ex">Building</span> DAG of jobs...</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="ex">Nothing</span> to be done.</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="ex">Complete</span> log: /home/zenmaster/data/yeast/.snakemake/log/2021-04-23T172441.519500.snakemake.log</span></code></pre>
</div>
<p>In normal operation, Snakemake only runs a job if:</p>
<ol style="list-style-type: decimal">
<li>A target file you explicitly requested to make is missing</li>
<li>An intermediate file is missing and it is needed in the process of
making a target file</li>
<li>Snakemake can see an input file which is newer than an output
file</li>
<li>A rule definition or configuration has changed since the output file
was created</li>
</ol>
<p>The last of these relies on a ledger that Snakemake saves into the
<code>.snakemake</code> directory.</p>
<p>Let’s demonstrate each of these in turn, by altering some files and
re-running Snakemake without the <code>-F</code> option.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> rm <span class="at">-rv</span> kallisto.temp33_1</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-j1</span> <span class="at">-p</span> kallisto.temp33_1/abundance.h5</span></code></pre>
</div>
<p>This just re-runs <em>kallisto_quant</em> - the final step.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> rm <span class="at">-v</span> trimmed/temp33_<span class="pp">*</span>.fq</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-j1</span> <span class="at">-p</span> kallisto.temp33_1/abundance.h5</span></code></pre>
</div>
<p>“Nothing to be done” - some intermediate output is missing but
Snakemake already has the file you are telling it to make, so it doesn’t
worry.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> touch transcriptome/<span class="pp">*</span>.fa.gz</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-j1</span> <span class="at">-p</span> kallisto.temp33_1/abundance.h5</span></code></pre>
</div>
<p>The <code>touch</code> command is a standard Linux command which sets
the timestamp of the file, so now the transcriptome looks to Snakemake
as if it was just modified.</p>
<p>Snakemake sees that one of the input files used in the process of
producing <code>kallisto.temp33_1/abundance.h5</code> is newer than the
existing output file, so it needs to run the <em>kallisto index</em> and
<em>kallisto quant</em> steps again. Of course, the <em>kallisto
quant</em> step needs the trimmed reads which we deleted earlier, so now
the trimming step is re-run also.</p>
</section><section id="explicitly-telling-snakemake-what-to-re-run"><h2 class="section-heading">Explicitly telling Snakemake what to re-run<a class="anchor" aria-label="anchor" href="#explicitly-telling-snakemake-what-to-re-run"></a>
</h2>
<hr class="half-width">
<p>The default timestamp-based logic is really useful when you want
to:</p>
<ol style="list-style-type: decimal">
<li>Change or add some inputs to an existing analysis without
re-processing everything</li>
<li>Continue running a workflow that failed part-way</li>
</ol>
<p>In most cases you can also rely on Snakemake to detect when you have
edited a rule, but sometimes you need to be explicit, for example if you
have updated an external script or changed a setting which Snakemake
doesn’t see.</p>
<p>The <code>-R</code> flag allows you to explicitly tell Snakemake that
a rule has changed and that all outputs from that rule need to be
re-evaluated.</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-j1</span> <span class="at">-R</span> trimreads <span class="at">-p</span> kallisto.temp33_1/abundance.h5</span></code></pre>
</div>
<div id="note-on--r" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="note-on--r" class="callout-inner">
<h3 class="callout-title">Note on<code>-R</code><a class="anchor" aria-label="anchor" href="#note-on--r"></a>
</h3>
<div class="callout-content">
<p>Due to a quirk of the way Snakemake parses command-line options, you
need to make sure there are options after the <code>-R ...</code>,
before the list of target outputs. If you don’t do this, Snakemake will
think that the target files are instead items to add to the
<code>-R</code> list, and then when building the DAG it will just try to
run the default rule.</p>
<p>The easiest way is to put the <code>-p</code> flag before the target
outputs. Then you can list multiple rules to re-run, and also multiple
targets, and Snakemake can tell which is which.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-j1</span> <span class="at">-R</span> trimreads kallisto_index <span class="at">-p</span> kallisto.temp33_1/abundance.h5 kallisto.temp33_2/abundance.h5</span></code></pre>
</div>
<p>The reason for using the <code>-p</code> flag here is because you
generally always want this option.</p>
</div>
</div>
</div>
<p>The <code>-f</code> flag specifies that the target outputs named on
the command line should always be regenerated, so you can use this to
explicitly re-make specific files.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-j1</span> <span class="at">-f</span> <span class="at">-p</span> kallisto.temp33_1/abundance.h5</span></code></pre>
</div>
<p>This always re-runs <em>kallisto_quant</em>, regardless of whether
the output file is there already. For all intermediate outputs,
Snakemake applies the default timestamp-based logic. Contrast with
<code>-F</code> which runs the entire DAG every time.</p>
</section><section id="visualising-the-dag"><h2 class="section-heading">Visualising the DAG<a class="anchor" aria-label="anchor" href="#visualising-the-dag"></a>
</h2>
<hr class="half-width">
<p>Snakemake can draw a picture of the DAG for you, if you run it like
this:</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-f</span> <span class="at">--dag</span> kallisto.etoh60_1/abundance.h5 <span class="kw">|</span> <span class="ex">gm</span> display <span class="at">-</span></span></code></pre>
</div>
<p>Using the <code>--dag</code> option implicitly activates the
<code>-n</code> (dry-run) option so that Snakemake will not actually run
any jobs, it will just print the DAG and stop. Snakemake prints the DAG
in a text format so we use the <code>gm</code> command to make this into
a picture and show it on the screen.</p>
<div id="note-on-gm-display" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="note-on-gm-display" class="callout-inner">
<h3 class="callout-title">Note on<code>gm display</code><a class="anchor" aria-label="anchor" href="#note-on-gm-display"></a>
</h3>
<div class="callout-content">
<p>The <code>gm</code> command is provided by the <a href="https://www.graphicsmagick.org/" class="external-link">GraphicsMagick toolkit</a>. On
systems where <code>gm</code> will not display an image directly, you
may instead save it to a PNG file. You will need the <code>dot</code>
program from the <a href="https://graphviz.org/" class="external-link">GraphViz package</a>
installed.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-f</span> <span class="at">--dag</span> kallisto.etoh60_1/abundance.h5 <span class="kw">|</span> <span class="ex">dot</span> <span class="at">-Tpng</span> <span class="op">&gt;</span> dag.png</span></code></pre>
</div>
</div>
</div>
</div>
<figure><img src="fig/dag_2.png" alt="A DAG for the partial workflow with four boxes, representing two trimreads jobs and a kallisto_index job, then a kallisto_quant job receiving input from the previous three, The boxes for the kallisto_index and trimreads jobs are dotted, but the kallisto_quant box is solid." class="figure mx-auto d-block"></figure><p>The boxes drawn with dotted lines indicate steps that are not to be
run, as the output files are already present and newer than the input
files.</p>
<div id="visualising-the-effect-of-the--r-and--f-flags" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="visualising-the-effect-of-the--r-and--f-flags" class="callout-inner">
<h3 class="callout-title">Visualising the effect of the<code>-R</code>and<code>-f</code>flags<a class="anchor" aria-label="anchor" href="#visualising-the-effect-of-the--r-and--f-flags"></a>
</h3>
<div class="callout-content">
<p>Run <em>kallisto_quant</em> on the first of the
<strong>etoh60</strong> samples, then use the <code>--dag</code> option
as shown above to check:</p>
<ol style="list-style-type: decimal">
<li><p>How many jobs will run if you ask again to create this output
with no <code>-f</code>, <code>-F</code> or <code>-R</code>
options?</p></li>
<li><p>How many if you use the <code>-f</code> option?</p></li>
<li><p>How many if you use the <code>-R trimreads</code>
option?</p></li>
<li><p>How many if you edit the Snakefile so that the
<code>qual_threshold</code> for <code>trimreads</code> is “22”, rather
than “20”?</p></li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">Show me the solution</h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<p>This is a way to make the Kallisto result in the first place:</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-j1</span> <span class="at">-p</span> kallisto.etoh60_1/abundance.h5</span></code></pre>
</div>
<ol style="list-style-type: decimal">
<li>This command should show four boxes, but all are dotted so no jobs
are actually to be run.</li>
</ol>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">--dag</span> kallisto.etoh60_1/abundance.h5 <span class="kw">|</span> <span class="ex">gm</span> display <span class="at">-</span></span></code></pre>
</div>
<ol start="2" style="list-style-type: decimal">
<li><p>The <code>-f</code> flag re-runs only the job to create the
output file, so in this case one box is solid, and only that job will
run.</p></li>
<li><p>With <code>-R trimreads</code>, the two <em>trimreads</em> jobs
will re-run, and Snakemake sees that this also requires re-running
<em>kallisto_quant</em>, so the answer is 3.</p></li>
</ol>
<p>If you see a message like the one below, it’s because you need to put
an option after <code>trimreads</code> or else Snakemake gets confused
about what are parameters of <code>-R</code>, and what things are
targets.</p>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>WorkflowError:
Target rules may not contain wildcards.</code></pre>
</div>
<ol start="4" style="list-style-type: decimal">
<li>Editing the Snakefile has the same effect as forcing the
<em>trimreads</em> rule to re-run, so again there will be three jobs to
be run from the DAG.</li>
</ol>
<p>With older versions of Snakemake this would not be auto-detected, and
in fact you can see this behaviour if you remove the hidden
<code>.snakemake</code> directory. Now Snakemake has no memory of the
rule change so it will not re-run any jobs unless explicitly told
to.</p>
</div>
</div>
</div>
</div>
<div id="removing-files-to-trigger-reprocessing" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="removing-files-to-trigger-reprocessing" class="callout-inner">
<h3 class="callout-title">Removing files to trigger reprocessing<a class="anchor" aria-label="anchor" href="#removing-files-to-trigger-reprocessing"></a>
</h3>
<div class="callout-content">
<p>In general, getting Snakemake to re-run things by removing files is a
bad idea, because it’s easy to forget about intermediate files that
actually contain stale results and need to be updated. Using the
<code>-R</code> flag is simpler and more reliable. If in doubt, and if
it will not be too time consuming, keep it simple and just use
<code>-F</code> to run the whole workflow from scratch.</p>
<p>For the opposite case where you want to avoid re-running particular
steps, see the <code>‑‑touch</code> option of Snakemake mentioned <a href="12-cleaning_up.html">later in the course.</a></p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>A <strong>job</strong> in Snakemake is a <strong>rule</strong> plus
<strong>wildcard values</strong> (determined by working back from the
requested output)</li>
<li>Snakemake plans its work by arranging all the jobs into a
<strong>DAG</strong> (directed acyclic graph)</li>
<li>If output files already exist, Snakemake can skip parts of the
DAG</li>
<li>Snakemake compares file timestamps and a log of previous runs to
determine what need regenerating</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-05-expansion"><p>Content from <a href="05-expansion.html">Processing lists of inputs</a></p>
<hr>
<p>Last updated on 2024-02-23 |
        
        <a href="https://github.com/carpentries-incubator/snakemake-novice-bioinformatics/edit/main/episodes/05-expansion.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do I process multiple files at once?</li>
<li>How do I define a default set of outputs for my Snakefile?</li>
<li>How do I make Snakemake detect what to process?</li>
<li>How do I combine multiple files together?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Use Snakemake to process all our samples at once</li>
<li>Make a summary of all read counts</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<p><em>For reference, <a href="files/ep03.Snakefile">this is the
Snakefile</a> you should have to start the episode. We didn’t modify it
during the last episode.</em></p>
<section id="defining-a-list-of-samples-to-process"><h2 class="section-heading">Defining a list of samples to process<a class="anchor" aria-label="anchor" href="#defining-a-list-of-samples-to-process"></a>
</h2>
<hr class="half-width">
<p>So far, we’ve told Snakemake what output files to generate by giving
the names of the desired files on the command line. Often you want
Snakemake to process all the available samples. How can we do this?</p>
<p>The <code>yeast/reads</code> directory contains results from three
conditions: <code>ref</code>, <code>etoh60</code> and
<code>temp33</code>. There are three replicates for each condition.
There is minor inconsistency in the naming convention, as the
<code>ref</code> files do not have an underscore before the replicate
number. Consistent naming is important for Snakemake, so let’s fix up
the names before we go any further.</p>
<p>A good way to do this is by making symlinks, because that way you
don’t lose sight of the original file names.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> mv reads original_reads</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> mkdir reads</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> cd reads</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> ln <span class="at">-s</span> ../original_reads/<span class="pp">*</span> .</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> rename <span class="at">-v</span> <span class="at">-s</span> ref ref_ <span class="pp">*</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> cd ..</span></code></pre>
</div>
<div id="file-renaming" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="file-renaming" class="callout-inner">
<h3 class="callout-title">File renaming<a class="anchor" aria-label="anchor" href="#file-renaming"></a>
</h3>
<div class="callout-content">
<p>The <strong>rename</strong> command here is <a href="https://plasmasturm.org/code/rename/" class="external-link">the one provided by
Bioconda</a>. Other Linux systems may have a different rename command
installed by default.</p>
</div>
</div>
</div>
<p>Having harmonized the file names we’ll tell Snakemake about our
conditions and replicates. To do this, we can define some lists as
Snakemake <strong>global variables</strong>.</p>
<p>Global variables should be added before the rules in the
Snakefile.</p>
<pre class="source"><code># Input conditions and replicates to process
CONDITIONS = ["ref", "etoh60", "temp33"]
REPLICATES = ["1", "2", "3"]</code></pre>
<ul>
<li>Unlike with variables in shell scripts, we can put spaces around the
<code>=</code> sign, but they are not mandatory.</li>
<li>The lists of quoted strings are enclosed in square brackets and
comma-separated. If you know any Python you’ll recognise this as Python
list syntax.</li>
<li>A good convention is to use capitalized names for these variables,
but this is not mandatory.</li>
<li>Although these are referred to as variables, you can’t actually
change the values once the workflow is running, so lists defined this
way are more like constants.</li>
</ul></section><section id="using-a-snakemake-rule-to-define-a-batch-of-outputs"><h2 class="section-heading">Using a Snakemake rule to define a batch of outputs<a class="anchor" aria-label="anchor" href="#using-a-snakemake-rule-to-define-a-batch-of-outputs"></a>
</h2>
<hr class="half-width">
<p>We’ll add another rule to our Snakefile. This special target rule
will have an <code>input</code> section but no <code>output</code> or
<code>shell</code> sections (yet).</p>
<pre class="source"><code>rule all_counts:
    input: expand("trimmed.{cond}_{rep}_1.fq.count", cond=CONDITIONS, rep=REPLICATES)</code></pre>
<p>The <code>expand(...)</code> function in this rule generates a list
of filenames, by taking the first thing in the parentheses as a template
and replacing <a href="https://www.r-project.org" class="external-link">cond</a> with all the <code>CONDITIONS</code>
and <code>{rep}</code> with all the <code>REPLICATES</code>. Since there
are 3 of each, this will yield 9 combinations - ie. 9 files we want to
make.</p>
<p>This list goes into the <em>input</em> section of the rule. You might
think that since these filenames are outputs that we want from our
workflow they should go into the <em>output</em> section. However,
remember that <em>output</em>s of a rule are things the rule can make
itself, and this rule doesn’t actually make anything. It’s just a
placeholder for a bunch of filenames.</p>
<p>We now tell Snakemake to make all these files by using the <em>target
rule name</em> on the command line:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-j1</span> <span class="at">-p</span> all_counts</span></code></pre>
</div>
<p>Here, Snakemake sees that <em>all_counts</em> is the name of a rule
in the Snakefile, so rather than trying to make a file literally named
<code>all_counts</code> it looks at all the input files for the rule and
tries to make them. In this case, all of the inputs to
<em>all_counts</em> can be made by the <em>countreads</em> rule, and all
of the inputs for those jobs are made by <em>trimreads</em>. The
resulting workflow is the same as if we had typed out all 9 of the
filenames on the command line.</p>
<p>If you don’t specify a target rule name or any file names on the
command line when running Snakemake, the default is to use <strong>the
first rule</strong> in the Snakefile as the target. So if
<em>all_counts</em> is defined at the top, before the other rules, you
can simply say:</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-j1</span> <span class="at">-p</span></span></code></pre>
</div>
<div id="rules-as-targets" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="rules-as-targets" class="callout-inner">
<h3 class="callout-title">Rules as targets<a class="anchor" aria-label="anchor" href="#rules-as-targets"></a>
</h3>
<div class="callout-content">
<p>Giving the name of a rule to Snakemake on the command line only works
when that rule has <em>no wildcards</em> in the outputs, because
Snakemake has no way to know what the desired wildcards might be. You
will see the error “Target rules may not contain wildcards.” This can
also happen when you don’t supply any explicit targets on the command
line at all, and Snakemake tries to run the first rule defined in the
Snakefile.</p>
</div>
</div>
</div>
<div id="counting-all-the-reads" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="counting-all-the-reads" class="callout-inner">
<h3 class="callout-title">Counting all the reads<a class="anchor" aria-label="anchor" href="#counting-all-the-reads"></a>
</h3>
<div class="callout-content">
<p>Check that the <em>all_counts</em> rule is working. Now adapt the
Snakefile so that it makes all the counts for both of the pairs of reads
(<code>_1.fq</code> and <code>_2.fq</code>), and also for both trimmed
and untrimmed versions of the files. So you should end up with 36 count
files in total.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">Show me the solution</h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>This will work.</p>
<pre class="source"><code># Input conditions and replicates to process
CONDITIONS = ["ref", "etoh60", "temp33"]
REPLICATES = ["1", "2", "3"]
READ_ENDS  = ["1", "2"]
COUNT_DIR  = ["reads", "trimmed"]

# Rule to make all counts at once
rule all_counts:
    input: expand("{indir}.{cond}_{rep}_{end}.fq.count", indir=COUNT_DIR, cond=CONDITIONS, rep=REPLICATES, end=READ_ENDS)</code></pre>
<p>Alternatively you can put the lists directly into the
<code>expand()</code> function rather than declaring more variables. To
aid readability of the code it’s also possible to split the function
over more than one line, but note that this only works if you put a
newline after the <code>input:</code> line too.</p>
<pre class="source"><code># Input conditions and replicates to process
CONDITIONS = ["ref", "etoh60", "temp33"]
REPLICATES = ["1", "2", "3"]

# Rule to make all counts at once
rule all_counts:
    input:
        expand( "{indir}.{cond}_{rep}_{end}.fq.count", indir = ["reads", "trimmed"],
                                                       cond  = CONDITIONS,
                                                       rep   = REPLICATES,
                                                       end   = ["1", "2"] )</code></pre>
</div>
</div>
</div>
</div>
</section><section id="rules-that-combine-multiple-inputs"><h2 class="section-heading">Rules that combine multiple inputs<a class="anchor" aria-label="anchor" href="#rules-that-combine-multiple-inputs"></a>
</h2>
<hr class="half-width">
<p>Our <em>all_counts</em> rule is a rule which takes a list of input
files. The length of that list is not fixed by the rule, but can change
based on <code>CONDITIONS</code> and <code>REPLICATES</code>. If we want
to perform some combining operation on the whole list of files, we can
add <code>output</code> and <code>shell</code> sections to this
rule.</p>
<p>In typical bioinformatics workflows, the final steps will combine all
the results together into some big report. For our final workflow we’ll
be doing this with <em>MultiQC</em>, but as a simple first example,
let’s just concatenate all the count files. In the shell this would
be:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> cat file1.count file2.count file3.count ... <span class="op">&gt;</span> all_counts.txt</span></code></pre>
</div>
<p>In the Snakemake rule we just say:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="ex">shell:</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="st">"cat {input} &gt; {output}"</span></span></code></pre>
</div>
<p>Within a rule definition you can combine named inputs and list inputs
- any named input can be list of files rather than just a single file.
When you use the <code>{input.name}</code> placeholder in the shell
command it will expand to the full list.</p>
<div id="combining-the-inputs-of-the-all_counts-rule" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="combining-the-inputs-of-the-all_counts-rule" class="callout-inner">
<h3 class="callout-title">Combining the inputs of the<em>all_counts</em>rule<a class="anchor" aria-label="anchor" href="#combining-the-inputs-of-the-all_counts-rule"></a>
</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>Make it so that the <em>all_counts</em> rule concatenates all the
count files into a single output file named
<code>all_counts_concatenated.txt</code>.</li>
<li>Adapt the rule further so that there are two outputs named
<code>trimmed_counts_concatenated.txt</code> and
<code>untrimmed_counts_concatenated.txt</code>, and the respective
counts go into each.</li>
</ol>
<p><em>Hint: you can put two commands into the <code>shell</code> part,
separated by a semicolon <code>;</code>.</em></p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">Show me the solution</h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<pre class="source"><code>rule all_counts:
    input:
        expand( "{indir}.{cond}_{rep}_{end}.fq.count", indir = ["reads", "trimmed"],
                                                       cond  = CONDITIONS,
                                                       rep   = REPLICATES,
                                                       end   = ["1", "2"] )
    output:
        "all_counts_concatenated.txt"
    shell:
        "cat {input} &gt; {output}"</code></pre>
<pre class="source"><code>rule all_counts:
    input:
        untrimmed = expand( "reads.{cond}_{rep}_{end}.fq.count",   cond  = CONDITIONS,
                                                                   rep   = REPLICATES,
                                                                   end   = ["1", "2"] ),
        trimmed   = expand( "trimmed.{cond}_{rep}_{end}.fq.count", cond  = CONDITIONS,
                                                                   rep   = REPLICATES,
                                                                   end   = ["1", "2"] ),
    output:
        untrimmed = "untrimmed_counts_concatenated.txt",
        trimmed   = "trimmed_counts_concatenated.txt",
    shell:
        "cat {input.untrimmed} &gt; {output.untrimmed} ; cat {input.trimmed} &gt; {output.trimmed}"</code></pre>
<p>To run either version of the rule:</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-j1</span> <span class="at">-p</span> all_counts</span></code></pre>
</div>
</div>
</div>
</div>
</div>
</section><section id="dynamically-determining-the-inputs"><h2 class="section-heading">Dynamically determining the inputs<a class="anchor" aria-label="anchor" href="#dynamically-determining-the-inputs"></a>
</h2>
<hr class="half-width">
<p>In the shell we can match multiple filenames with <strong>glob
patterns</strong> like <code>original_reads/*</code> or
<code>reads/ref?_?.fq</code>. Snakemake allows you to do something
similar with the <code>glob_wildcards()</code> function, so we’ll use
this in our Snakefile.</p>
<pre class="source"><code>CONDITIONS = glob_wildcards("reads/{condition}_1_1.fq").condition
print("Conditions are: ", CONDITIONS)</code></pre>
<p>Here, the list of conditions is captured from the files seen in the
reads directory. The pattern used in <code>glob_wildcards()</code> looks
much like the input and output parts of rules, with a wildcard in {curly
brackets}, but here the pattern is being used to search for matching
files. We’re only looking for read 1 of replicate 1 so this will return
just three matches</p>
<p>Rather than getting the full file names, the function yields the
values of the wildcard, which we can assign directly to a list. The
<code>print()</code> statement will print out the value of
<code>CONDITIONS</code> when the Snakefile is run (including dry-run
mode, activated with the <code>-n</code> option as mentioned in <a href="02-placeholders.html">Episode 2</a>), and reassures us that the
list really is the same as before.</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-j1</span> <span class="at">-F</span> <span class="at">-n</span> <span class="at">-p</span> all_counts</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="ex">Conditions</span> are:  [<span class="st">'etoh60'</span>, <span class="st">'temp33'</span>, <span class="st">'ref'</span>]</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="ex">Building</span> DAG of jobs...</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="ex">Job</span> counts:</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>    <span class="ex">count</span>  jobs</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>    <span class="ex">1</span>      all_counts</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>    <span class="ex">36</span>     countreads</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>    <span class="ex">18</span>     trimreads</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>    <span class="ex">55</span></span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a><span class="ex">...</span></span></code></pre>
</div>
<p>Using <code>glob_wildcards()</code> gets a little more tricky when
you need a more complex match. To refine the match we can quickly test
out results by activating the Python interpreter. This saves editing the
Snakefile and running Snakemake just to see what
<code>glob_wildcards()</code> will match.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>$ python3</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="im">from</span> snakemake.io <span class="im">import</span> glob_wildcards, expand</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> glob_wildcards(<span class="st">"reads/</span><span class="sc">{condition}</span><span class="st">_1_1.fq"</span>)</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>Wildcards(condition<span class="op">=</span>[<span class="st">'etoh60'</span>, <span class="st">'temp33'</span>, <span class="st">'ref'</span>])</span></code></pre>
</div>
<p>This is the result we got before. So far, so good.</p>
<div id="the-python-interpreter" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="the-python-interpreter" class="callout-inner">
<h3 class="callout-title">The Python interpreter<a class="anchor" aria-label="anchor" href="#the-python-interpreter"></a>
</h3>
<div class="callout-content">
<p>The Python interpreter is like a special shell for Python commands,
and is a familiar to anyone who has learned Python. Because Snakemake
functions are actually Python functions they can be run from the
interpreter after being imported.</p>
<p>Note that <code>&gt;&gt;&gt;</code> is the Python interpreter prompt,
and not part of the command to be typed. You can exit back to the
regular shell prompt by typing <code>exit()</code>. If you do exit and
then restart the interpreter, you will need to repeat the
<code>import</code> line.</p>
</div>
</div>
</div>
<div id="globbing-the-list-of-samples" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="globbing-the-list-of-samples" class="callout-inner">
<h3 class="callout-title">‘Globbing’ the list of samples<a class="anchor" aria-label="anchor" href="#globbing-the-list-of-samples"></a>
</h3>
<div class="callout-content">
<p>Staying in the Python interpreter, use <code>glob_wildcards()</code>
to list the names of all nine samples, that is the three replicates of
each condition.</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">Show me the solution</h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> glob_wildcards(<span class="st">"reads/</span><span class="sc">{sample}</span><span class="st">_1.fq"</span>)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>Wildcards(sample<span class="op">=</span>[<span class="st">'temp33_3'</span>, <span class="st">'temp33_2'</span>, <span class="st">'etoh60_1'</span>, <span class="st">'etoh60_3'</span>, <span class="st">'ref_2'</span>, <span class="st">'temp33_1'</span>, <span class="st">'etoh60_2'</span>, <span class="st">'ref_1'</span>, <span class="st">'ref_3'</span>])</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="globbing-the-list-of-samples" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="globbing-the-list-of-samples" class="callout-inner">
<h3 class="callout-title">‘Globbing’ the list of samples<em>(continued)</em><a class="anchor" aria-label="anchor" href="#globbing-the-list-of-samples"></a>
</h3>
<div class="callout-content">
<p>Still in the Python interpreter, use the <code>expand()</code>
function in combination with the <code>glob_wildcards()</code> to make a
list of all the count files, and then all the <em>kallisto_quant</em>
output files that would be generated for all the nine samples.</p>
<p>Remember you can save the result of <code>glob_wildcards()</code> to
a variable - this works the same way in the Python interpreter as it
does in the Snakefile.</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4">Show me the solution</h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> SAMPLES <span class="op">=</span> glob_wildcards(<span class="st">"reads/</span><span class="sc">{sample}</span><span class="st">_1.fq"</span>).sample</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> expand(<span class="st">"</span><span class="sc">{indir}</span><span class="st">.</span><span class="sc">{sample}</span><span class="st">_</span><span class="sc">{end}</span><span class="st">.fq.count"</span>, indir<span class="op">=</span>[<span class="st">'reads'</span>, <span class="st">'trimmed'</span>], sample<span class="op">=</span>SAMPLES, end<span class="op">=</span>[<span class="st">"1"</span>,<span class="st">"2"</span>])</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>...</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> expand(<span class="st">"kallisto.</span><span class="sc">{sample}</span><span class="st">/</span><span class="sc">{outfile}</span><span class="st">"</span>, sample<span class="op">=</span>SAMPLES, outfile<span class="op">=</span>[<span class="st">'abundance.h5'</span>, <span class="st">'abundance.tsv'</span>, run_info.json<span class="st">'])</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="glob-with-multiple-wildcards" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="glob-with-multiple-wildcards" class="callout-inner">
<h3 class="callout-title">Glob with multiple wildcards<a class="anchor" aria-label="anchor" href="#glob-with-multiple-wildcards"></a>
</h3>
<div class="callout-content">
<p>If there are two or more wildcards in the glob pattern, dealing with
the result becomes a little more tricky, but it does work.</p>
<p>Here is one way to re-combine two wildcards using
<code>expand()</code> and <code>zip</code> (demonstrated in the Python
interpreter as above).</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">PYTHON<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode python" tabindex="0"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>$ python3</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="im">from</span> snakemake.io <span class="im">import</span> glob_wildcards, expand</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> DOUBLE_MATCH <span class="op">=</span> glob_wildcards( <span class="st">"reads/</span><span class="sc">{condition}</span><span class="st">_</span><span class="sc">{samplenum}</span><span class="st">_1.fq"</span> )</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> DOUBLE_MATCH</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>Wildcards(condition<span class="op">=</span>[<span class="st">'temp33'</span>, <span class="st">'temp33'</span>, <span class="st">'etoh60'</span>, <span class="st">'etoh60'</span>, <span class="st">'ref'</span>, <span class="st">'temp33'</span>, <span class="st">'etoh60'</span>, <span class="st">'ref'</span>, <span class="st">'ref'</span>],</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>          samplenum<span class="op">=</span>[<span class="st">'3'</span>, <span class="st">'2'</span>, <span class="st">'1'</span>, <span class="st">'3'</span>, <span class="st">'2'</span>, <span class="st">'1'</span>, <span class="st">'2'</span>, <span class="st">'1'</span>, <span class="st">'3'</span>])</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> SAMPLES <span class="op">=</span> expand(</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>...     <span class="st">"</span><span class="sc">{condition}</span><span class="st">_</span><span class="sc">{samplenum}</span><span class="st">"</span>,</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>...     <span class="bu">zip</span>,</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>...     condition <span class="op">=</span> DOUBLE_MATCH.condition,</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>...     samplenum <span class="op">=</span> DOUBLE_MATCH.samplenum )</span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> SAMPLES</span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>[<span class="st">'temp33_3'</span>, <span class="st">'temp33_2'</span>, <span class="st">'etoh60_1'</span>, <span class="st">'etoh60_3'</span>, <span class="st">'ref_2'</span>, <span class="st">'temp33_1'</span>, <span class="st">'etoh60_2'</span>, <span class="st">'ref_1'</span>, <span class="st">'ref_3'</span>]</span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a><span class="op">&gt;&gt;&gt;</span> <span class="bu">sorted</span>(SAMPLES)</span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a>[<span class="st">'etoh60_1'</span>, <span class="st">'etoh60_2'</span>, <span class="st">'etoh60_3'</span>, <span class="st">'ref_1'</span>, <span class="st">'ref_2'</span>, <span class="st">'ref_3'</span>, <span class="st">'temp33_1'</span>, <span class="st">'temp33_2'</span>, <span class="st">'temp33_3'</span>]</span></code></pre>
</div>
</div>
</div>
</div>
<div id="rules-that-make-multiple-outputs" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="rules-that-make-multiple-outputs" class="callout-inner">
<h3 class="callout-title">Rules that make multiple outputs<a class="anchor" aria-label="anchor" href="#rules-that-make-multiple-outputs"></a>
</h3>
<div class="callout-content">
<p>If we can have rules that combine lists of files, can we do the
opposite and have a rule that produces a list of outputs?</p>
<p>The answer is yes, but the situation is not completely symmetrical.
Remember that Snakemake works out the full list of input and output
files to every job <em>before</em> it runs a single job in the workflow.
For a combining step, Snakemake will know how many samples/replicates
are being combined from the start. For a splitting step, it may or may
not be possible to predict the number of output files in advance. For
cases where you really do need to handle a dynamic list of outputs,
Snakemake has things called <a href="https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#data-dependent-conditional-execution" class="external-link">checkpoint
rules</a>.</p>
<p>In practise these are very rarely needed, so we’ll not be covering
them here in the course.</p>
</div>
</div>
</div>
<p><em>For reference, <a href="files/ep05.Snakefile">this is a
Snakefile</a> incorporating the changes made in this episode.</em></p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Rename your input files if necessary to maintain consistent
naming</li>
<li>List the things you want to proces as global variables, or discover
input files with <code>glob_wildcards()</code>
</li>
<li>Use the <code>expand()</code> function to generate lists of
filenames you want to combine</li>
<li>These functions can be tested in the Python interpreter</li>
<li>Any <code>{input}</code> to a rule can be a variable-length list,
but variable lists of outputs are trickier and rarely needed</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-06-awkward_programs"><p>Content from <a href="06-awkward_programs.html">Handling awkward programs</a></p>
<hr>
<p>Last updated on 2024-03-01 |
        
        <a href="https://github.com/carpentries-incubator/snakemake-novice-bioinformatics/edit/main/episodes/06-awkward_programs.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do I handle tools which don’t let me specify output file
names?</li>
<li>How do I define a rule where the output is a directory?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Add the FastQC tool to the pipeline</li>
<li>Understand some different choices available when defining a
rule</li>
<li>Learn about the <code>directory()</code> function</li>
<li>Add multiple command lines to the <code>shell</code> section of a
rule</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<p><em>For reference, <a href="files/ep05.Snakefile">this is the
Snakefile</a> you should have to start the episode.</em></p>
<section id="introducing-fastqc"><h2 class="section-heading">Introducing FastQC<a class="anchor" aria-label="anchor" href="#introducing-fastqc"></a>
</h2>
<hr class="half-width">
<p>FastQC is a popular tool for scanning FASTQ files and producing a
selection of quality plots. It’s “fast” in that it runs quickly, and
also in that you can normally use the default options so there is no
configuration needed.</p>
<figure><img src="fig/fastqc.png" alt="Screenshot of a typical FastQC report specifically showing the per-base quality box-and-whisker plot. This is one of eleven views that are shown as a selectable list in the application window. The plot itself shows vertical yellow bars that get increasingly taller and lower from left to right, indicating how the base quality in these short reads deteriorates as the run progresses. There is a red X icon next to this plot, while other views listed have green ticks or yellow exclamation point icons." class="figure mx-auto d-block"></figure><p>The program can be run interactively or in batch mode, where it saves
out results as an HTML file plus a ZIP file. We’ll obviously need to use
the batch mode to include it as part of our workflow, and we’ll see
shortly that this presents a minor problem.</p>
<p>In general, real bioinformatics programs like FastQC can have quirks
like:</p>
<ul>
<li>Not allowing you to specify <em>input</em> or <em>output</em> file
names</li>
<li>Outputting a whole directory of files</li>
<li>Creating temporary files in various places</li>
<li>Not supporting robust error handling</li>
</ul>
<p>With a little care we can handle all these problems while still
keeping our workflow tidy.</p>
</section><section id="adding-a-fastqc-rule"><h2 class="section-heading">Adding a FastQC rule<a class="anchor" aria-label="anchor" href="#adding-a-fastqc-rule"></a>
</h2>
<hr class="half-width">
<p>We’ll first run FastQC on a single input file and see what is
produced.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> fastqc reads/ref_1_1.fq</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> ls <span class="at">-ltr</span> reads</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="ex">...</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="ex">-rw-r--r--</span> 1 zenmaster  users   464852 Jun  9 14:31 ref_1_1_fastqc.zip</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="ex">-rw-r--r--</span> 1 zenmaster  users   654810 Jun  9 14:31 ref_1_1_fastqc.html</span></code></pre>
</div>
<p>It’s possible to supply multiple input files, but the resulting
output is exactly the same as if processing the files one at a time. Two
files are produced for each FASTQ file, and these files appear in the
same directory as the input file. The <code>fastqc</code> command does
not let us choose the output filenames, but we can set an alternative
output directory.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> fastqc <span class="at">--help</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="ex">...</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>   <span class="ex">-o</span> <span class="at">--outdir</span>     Create all output files in the specified output directory.</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>                   <span class="ex">Please</span> note that this directory must exist as the program</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>                   <span class="ex">will</span> not create it.  If this option is not set then the</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>                   <span class="ex">output</span> file for each sequence file is created in the same</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>                   <span class="ex">directory</span> as the sequence file which was processed.</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="ex">...</span></span></code></pre>
</div>
<p>For the <code>countreads</code> rule we wrote earlier (see episode
3), we chose our preferred output file name pattern so as to allow us to
effectively link rules. This gives us a rule that can count both
<em>trimmed</em> and <em>untrimmed</em> reads.</p>
<pre class="source"><code># Our existing countreads rule, for reference...
rule countreads:
    output: "{indir}.{myfile}.fq.count"
    input:  "{indir}/{myfile}.fq"
    shell:
        "echo $(( $(wc -l &lt;{input}) / 4 )) &gt; {output}"</code></pre>
<p>To do the same with FastQC, to report on both the trimmed and
untrimmed reads, we have various options:</p>
<ol style="list-style-type: decimal">
<li>Work with the default file names produced by FastQC and leave the
reports in the same directory with the FASTQ files.</li>
<li>Make the outputs in a new directory named, eg.
“reads.fastqc.ref_1_1/” (similar to what we did with Kallisto).</li>
<li>Do this, but tell Snakemake that the <em>directory itself</em> is
the output.</li>
<li>Force our preferred naming convention by renaming the FastQC output
files within the rule.</li>
</ol>
<p>We’ll try all four, and see where this gets us.</p>
<div class="section level3">
<h3 id="option-1-working-with-the-default-fastqc-output-files">Option 1: Working with the default FastQC output files<a class="anchor" aria-label="anchor" href="#option-1-working-with-the-default-fastqc-output-files"></a>
</h3>
<div id="adding-a-fastqc-rule-using-the-default-output-file-names" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="adding-a-fastqc-rule-using-the-default-output-file-names" class="callout-inner">
<h3 class="callout-title">Adding a FastQC rule using the default output
file names<a class="anchor" aria-label="anchor" href="#adding-a-fastqc-rule-using-the-default-output-file-names"></a>
</h3>
<div class="callout-content">
<p>Fill in the <code>???</code> to make a working rule for FastQC where
<code>indir</code> may be “reads” or “trimmed”. Do not change the shell
command or input pattern at all. Remember FastQC always makes two output
files, so add two named outputs.</p>
<pre class="source"><code>rule fastqc:
    output:
        ???
    input:  "{indir}/{myfile}.fq"
    shell:
        "fastqc {input}"</code></pre>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">Show me the solution</h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>Since the <code>shell</code> command is not to be changed, the output
names will be dictated by FastQC as we saw when running the command
directly in the terminal.</p>
<pre class="source"><code>rule fastqc:
    output:
        html = "{indir}/{myfile}_fastqc.html",
        zip  = "{indir}/{myfile}_fastqc.zip",
    input:  "{indir}/{myfile}.fq"
    shell:
        "fastqc {input}"</code></pre>
<p>This rule contains wildcards, so in order to run it you specify one
or more target output files:</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-j1</span> <span class="at">-p</span> reads/etoh60_1_1_fastqc.html reads/etoh60_1_2_fastqc.html</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<p>This rule is fine, but maybe we don’t want to put the reports in with
the sequences. As a general principle, when writing Snakemake rules, we
want to be in charge of output file names. FastQC lets us specify the
output directory, so we can use that…</p>
</div>
<div class="section level3">
<h3 id="option-2-using-the-default-output-filenames-in-a-new-directory">Option 2: Using the default output filenames in a new directory<a class="anchor" aria-label="anchor" href="#option-2-using-the-default-output-filenames-in-a-new-directory"></a>
</h3>
<div id="a-fastqc-rule-where-the-output-files-go-into-a-new-directory" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="a-fastqc-rule-where-the-output-files-go-into-a-new-directory" class="callout-inner">
<h3 class="callout-title">A FastQC rule where the output files go into a
new directory<a class="anchor" aria-label="anchor" href="#a-fastqc-rule-where-the-output-files-go-into-a-new-directory"></a>
</h3>
<div class="callout-content">
<p>Modify the rule so that the output files go into a new directory.
This will be very similar to the rule for <code>kallisto quant</code>
added in episode 3.</p>
<p>For example, when running on the file “trimmed/ref_1_1.fq” the
outputs should be</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="va">trimmed.fastqc.ref_1_1</span><span class="op">/</span><span class="va">ref_1_1_fastqc.html</span></span>
<span><span class="va">trimmed.fastqc.ref_1_1</span><span class="op">/</span><span class="va">ref_1_1_fastqc.zip</span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">Show me the solution</h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<p>This involves using the {myfile} wildcard twice and then constructing
the output directory name to place in the <code>-o</code> option to
fastqc.</p>
<pre class="source"><code>rule fastqc:
    output:
      html = "{indir}.fastqc.{myfile}/{myfile}_fastqc.html",
      zip  = "{indir}.fastqc.{myfile}/{myfile}_fastqc.zip",
    input: "{indir}/{myfile}.fq"
    shell:
        "fastqc -o {wildcards.indir}.fastqc.{wildcards.myfile} {input}"</code></pre>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="option-3-using-a-directory-output">Option 3: Using a <code>directory()</code> output<a class="anchor" aria-label="anchor" href="#option-3-using-a-directory-output"></a>
</h3>
<p>Our next option is to tell Snakemake not to worry about the
individual files at all, and just say that the output of the rule is a
whole directory. This makes the rule definition simpler.</p>
<p>We’ll amend the <em>fastqc</em> rule like so:</p>
<pre class="source"><code>rule fastqc:
    output: directory("{indir}.fastqc.{myfile}")
    input:  "{indir}/{myfile}.fq"
    shell:
        "fastqc -o {output} {input}"</code></pre>
<div id="only-use-directory-on-outputs-not-inputs" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="only-use-directory-on-outputs-not-inputs" class="callout-inner">
<h3 class="callout-title">Only use<code>directory()</code>on outputs,
not inputs<a class="anchor" aria-label="anchor" href="#only-use-directory-on-outputs-not-inputs"></a>
</h3>
<div class="callout-content">
<p>You only use the <code>directory()</code> declaration for outputs.
Any input to a rule may be a directory without the need for any special
syntax.</p>
</div>
</div>
</div>
<p>On running this, we get an error.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-j1</span> <span class="at">-p</span> reads.fastqc.ref_1_1</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="ex">...</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="ex">Specified</span> output directory <span class="st">'reads.fastqc.ref_1_1'</span> does not exist</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="ex">...</span></span></code></pre>
</div>
<p>This error is being printed by FastQC. FastQC requires that the
output directory must exist. (Other programs might insist that the
output directory does <em>not</em> already exist.) The error can be
rectified by making the directory explicitly in the <code>shell</code>
code.</p>
<pre class="source"><code>rule fastqc:
    output: directory("{indir}.fastqc.{myfile}")
    input:  "{indir}/{myfile}.fq"
    shell:
        "mkdir {output} ; fastqc -o {output} {input}"</code></pre>
<div id="the-rationale-for-making-output-directories" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="the-rationale-for-making-output-directories" class="callout-inner">
<h3 class="callout-title">The rationale for making output
directories<a class="anchor" aria-label="anchor" href="#the-rationale-for-making-output-directories"></a>
</h3>
<div class="callout-content">
<p>Remember that in most cases it is not necessary to manually create
directories because Snakemake will auto-create the directory for every
output file listed by a rule. When using a <code>directory()</code>
output, Snakemake will not create the directory itself but most
applications will make the directory for you. FastQC is an exception.
The best approach is to only add a <code>mkdir</code> command if you
first test the rule without it and see an error.</p>
</div>
</div>
</div>
<p>The modified rule works because the <code>shell</code> part of the
rule can contain a whole script, with multiple commands to be run. Above
we used a semicolon to split the commands on one line. For putting
multiple lines into a <code>shell</code> section there is a special
quoting syntax.</p>
<pre class="source"><code>rule fastqc:
    output: directory("{indir}.fastqc.{myfile}")
    input:  "{indir}/{myfile}.fq"
    shell:
        """mkdir {output}
           fastqc -o {output} {input}
        """</code></pre>
<p>The “triple quoting” syntax comes from Python. Not only does it allow
multiple lines to be added within the quotes but it also allows you to
embed both single and double quotes into the shell commands. For a
further discussion of string quoting and a way to disable the
interpretation of “backslash escapes” like <code>\n</code> and
<code>\t</code> see <a href="13-quoting.html">episode 13</a></p>
<p>This rule is also fine, but because the individual files are not
explicitly named as outputs we may have problems chaining later rules.
Also consider that some applications won’t give you any control at all
over the output filenames. The most powerful solution is to use shell
commands to move and/or rename the files to exactly the names you
want.</p>
</div>
<div class="section level3">
<h3 id="option-4-insisting-on-our-own-file-names">Option 4: Insisting on our own file names<a class="anchor" aria-label="anchor" href="#option-4-insisting-on-our-own-file-names"></a>
</h3>
<div id="fixing-fastqc-to-use-our-own-output-file-names" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="fixing-fastqc-to-use-our-own-output-file-names" class="callout-inner">
<h3 class="callout-title">Fixing FastQC to use our own output file
names<a class="anchor" aria-label="anchor" href="#fixing-fastqc-to-use-our-own-output-file-names"></a>
</h3>
<div class="callout-content">
<p>Complete the rule below so that the output filenames are correctly
produced. You will need to add extra commands to the <code>shell</code>
part after running <code>fastqc</code>. Do not alter the
<code>output</code> or <code>input</code> parts of the rule.</p>
<pre class="source"><code>rule fastqc:
    output:
        html = "{indir}.{myfile}_fastqc.html",
        zip  = "{indir}.{myfile}_fastqc.zip"
    input:  "{indir}/{myfile}.fq"
    shell:
        """???
        """</code></pre>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">Show me the solution</h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<p>This is one solution, using <code>-o .</code> to tell FastQC to
produce the files in the current directory, then explicitly renaming
them to match the declared rule outputs. Remember that, after Snakemake
runs all the commands in the <code>shell</code> block, it checks to see
that all the expected <code>output</code> files are present, but within
the block you can run any commands and make any files you like.</p>
<pre class="source"><code>rule fastqc:
    output:
        html = "{indir}.{myfile}_fastqc.html",
        zip  = "{indir}.{myfile}_fastqc.zip"
    input:  "{indir}/{myfile}.fq"
    shell:
        """fastqc -o . {input}
           mv {wildcards.myfile}_fastqc.html {output.html}
           mv {wildcards.myfile}_fastqc.zip  {output.zip}
        """</code></pre>
</div>
</div>
</div>
</div>
<div id="one-problem-with-making-intermediate-outputs" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="one-problem-with-making-intermediate-outputs" class="callout-inner">
<h3 class="callout-title">One problem with making intermediate
outputs<a class="anchor" aria-label="anchor" href="#one-problem-with-making-intermediate-outputs"></a>
</h3>
<div class="callout-content">
<p>There is actually a problem with the above solution which only starts
to matter when we allow Snakemake to run multiple jobs in parallel.
Right now we are always using <code>-j1</code>, but if we used, eg.
<code>-j2</code>, then potentially Snakemake may try to make
“reads.ref_1_1_fastqc.html” and “trimmed.ref_1_1_fastqc.html” in
parallel, and both instances would be trying to write to the same
temporary files at the same time. Snakemake has an elegant solution to
this, in the form of <code>shadow</code> rules, but we’re getting ahead
of ourselves. For now we’re running one job at a time, and this will
work.</p>
</div>
</div>
</div>
</div>
</section><section id="altering-the-kallisto-rule-to-have-a-directory-output"><h2 class="section-heading">Altering the Kallisto rule to have a <code>directory()</code>
output<a class="anchor" aria-label="anchor" href="#altering-the-kallisto-rule-to-have-a-directory-output"></a>
</h2>
<hr class="half-width">
<p>We saw above that the output of a rule can be a directory and saw the
<code>directory()</code> function which declares this. If you remember
the rule for <code>kallisto quant</code> you may be thinking that this
could have been written with the whole directory as the output, and you
would be right.</p>
<pre class="source"><code># Existing rule for kallisto_quant
rule kallisto_quant:
    output:
        h5   = "kallisto.{sample}/abundance.h5",
        tsv  = "kallisto.{sample}/abundance.tsv",
        json = "kallisto.{sample}/run_info.json",
    ...

# Alternative rule with directory() output
rule kallisto_quant:
    output:  directory("kallisto.{sample}")
    ...</code></pre>
<p>Make the change in your Snakefile now. In other workflows this might
not be the right approach but in this case it works fine and makes the
Snakefile neater. It will also make sense in the next chapter where we
add the remaining rules and finish the workflow.</p>
<p><em>For reference, <a href="files/ep06.Snakefile">this is a
Snakefile</a> incorporating the changes made in this episode.</em></p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Different bioinformatics tools will have different quirks</li>
<li>If programs limit your options for choosing input and output
filenames, you have several ways to deal with this</li>
<li>Use triple-quote syntax to make longer shell scripts with multiple
commands</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-07-finished-workflow"><p>Content from <a href="07-finished-workflow.html">Finishing the basic workflow</a></p>
<hr>
<p>Last updated on 2024-03-01 |
        
        <a href="https://github.com/carpentries-incubator/snakemake-novice-bioinformatics/edit/main/episodes/07-finished-workflow.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What does the full sample workflow look like?</li>
<li>How can we report some initial results from this analysis?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Finish the sample workflow to produce a final MultiQC report</li>
<li>See more ways to handle awkward software by adding extra setup shell
commands</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<p><em>For reference, <a href="files/ep06.Snakefile">this is the
Snakefile</a> you should have to start the episode.</em></p>
<p>We’ve seen how to link rules in a pipeline and how to merge all the
results at the final step. This is the basic pattern for many analysis
workflows. For simplicity, in episode 5, we just used the
<code>cat</code> command to combine all the <code>.count</code> files
but now we’ll use <em>MultiQC</em> to combine the results from Kallisto
and FastQC into a single report for all samples. We’ll also add in an
alternative quantification tool called <em>Salmon</em> and this will
complete the pipeline.</p>
<section id="the-full-workflow-we-are-constructing"><h2 class="section-heading">The full workflow we are constructing<a class="anchor" aria-label="anchor" href="#the-full-workflow-we-are-constructing"></a>
</h2>
<hr class="half-width">
<ul>
<li>
<strong>fastq_quality_trimmer</strong> is part of the FastX toolkit
and removes low-quality basecalls from the raw reads. <em>We first used
it in <a href="02-placeholders.html">episode 2</a>.</em>
</li>
<li>
<strong>FastQC</strong> calculates a variety of metrics on a FASTQ
file and produces an HTML report and a ZIP file. <em>We introduced this
in <a href="06-awkward_programs.html">episode 6</a>.</em>
</li>
<li>
<strong>Kallisto</strong> aligns the reads to a reference
transcriptome and produces a table of transcript abundance. <em>We first
used it in <a href="03-chaining_rules.html">episode 3</a>.</em>
</li>
<li>
<strong>Salmon</strong> is a alternative to Kallisto, using a
different alignment algorithm. <em>We’ve not used it yet.</em>
</li>
<li>
<strong>MultiQC</strong> combines the reports from various tools,
including FastQC, Kallisto, and Salmon, into a single HTML report over
all samples. This is by no means a full RNA-Seq analysis report but
today it completes our Snakemake pipeline.</li>
</ul>
<figure><img src="fig/overview.svg" alt="Summary of our full QC workflow with icons representing the steps listed above. The input data is also summarized, with 18 paired FASTQ files under yeast/reads, for the three repeats of all three conditions, as well as the transcriptome in gzipped FASTA format." class="figure mx-auto d-block"></figure><p>At this point we have everything we need, in terms of Snakemake
knowledge, to add the two remaining tools and complete the Snakefile. As
with FastQC, some quirks in the way the new tools work will need to be
accounted for.</p>
</section><section id="adding-salmon-as-an-alternative-to-kallisto"><h2 class="section-heading">Adding Salmon as an alternative to Kallisto<a class="anchor" aria-label="anchor" href="#adding-salmon-as-an-alternative-to-kallisto"></a>
</h2>
<hr class="half-width">
<p>At the end of the previous episode we modified the <em>kallisto</em>
rule by declaring that the output of the rule was a directory, rather
than explicitly listing the three files that Kallisto writes in that
directory. You should ensure that you are working with this new version
of the <em>kallisto</em> rule as it will be a template for adding a
<em>salmon</em> rule.</p>
<div id="adding-salmon-as-an-alternative-to-kallisto-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="adding-salmon-as-an-alternative-to-kallisto-1" class="callout-inner">
<h3 class="callout-title">Adding Salmon as an alternative to
Kallisto<a class="anchor" aria-label="anchor" href="#adding-salmon-as-an-alternative-to-kallisto-1"></a>
</h3>
<div class="callout-content">
<p>An alternative application for transcript quantification is Salmon.
The procedure is virtually identical, to Kallisto, having an indexing
step and a quantification step. Note that in real usage you are advised
to prepare and add decoy sequences to the transcriptome index, but for
the purposes of this tutorial we’ll just keep things as simple as
possible.</p>
<p>Based upon the following command templates:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> salmon index <span class="at">-t</span> <span class="op">&lt;</span>transcriptome as fasta<span class="op">&gt;</span> -i <span class="op">&lt;</span>index name<span class="op">&gt;</span> -k 31</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> salmon quant <span class="at">-i</span> <span class="op">&lt;</span>index name<span class="op">&gt;</span> -l A <span class="at">-1</span> <span class="op">&lt;</span>fastq1<span class="op">&gt;</span> -2 <span class="op">&lt;</span>fastq2<span class="op">&gt;</span> --validateMappings <span class="at">-o</span> <span class="op">&lt;</span>output path<span class="op">&gt;</span></span></code></pre>
</div>
<p>Add a pair of rules to index and align the reads with Salmon. Note
that:</p>
<ol style="list-style-type: decimal">
<li>Unlike Kallisto, the index produced by Salmon is a directory of
files, not a single file - so both of these new rules will produce a
directory as output.</li>
<li>As noted in the last episode, you only need the
<code>directory()</code> marker on the outputs of rules, not the
inputs.</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">Show me the solution</h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<pre class="source"><code>rule salmon_quant:
    output: directory("salmon.{sample}")
    input:
        index = "Saccharomyces_cerevisiae.R64-1-1.salmon_index",
        fq1   = "trimmed/{sample}_1.fq",
        fq2   = "trimmed/{sample}_2.fq",
    shell:
        "salmon quant -i {input.index} -l A -1 {input.fq1} -2 {input.fq2} --validateMappings -o {output}"

rule salmon_index:
    output:
        idx = directory("{strain}.salmon_index")
    input:
        fasta = "transcriptome/{strain}.cdna.all.fa.gz"
    shell:
        "salmon index -t {input.fasta} -i {output.idx} -k 31"</code></pre>
<p>If you copied the <em>kallisto_index</em> rule and logged the output
of <em>salmon_index</em> to a file this is fine. Just make sure when
copying and pasting that you change all the parts that need to
change!</p>
</div>
</div>
</div>
</div>
</section><section id="combining-the-outputs-with-multiqc"><h2 class="section-heading">Combining the outputs with MultiQC<a class="anchor" aria-label="anchor" href="#combining-the-outputs-with-multiqc"></a>
</h2>
<hr class="half-width">
<p>MultiQC scans for analysis report files in a given directory and all
subdirectories, then makes a report of everything it finds. It knows
about FastQC, Salmon and Kallisto outputs so we should be able to
compile a report on all these. To try this out and scan the current
directory, simply run:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> multiqc . <span class="at">-o</span> multiqc_out</span></code></pre>
</div>
<div id="adding-a-multiqc-rule" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="adding-a-multiqc-rule" class="callout-inner">
<h3 class="callout-title">Adding a MultiQC rule<a class="anchor" aria-label="anchor" href="#adding-a-multiqc-rule"></a>
</h3>
<div class="callout-content">
<p>Earlier, in episode 5, we made a basic summary-type rule called
<em>all_counts</em>. Now make a <em>multiqc</em> rule that gathers up
all the FastQC, Salmon and Kallisto reports.</p>
<p>Considerations:</p>
<ol style="list-style-type: decimal">
<li>Your rule is going to have several named inputs, and these inputs
will be lists of files generated with <code>expand()</code>
functions.</li>
<li>Ensure that both <em>kallisto_quant</em> and <em>salmon_quant</em>
are run on all 9 samples, that is all three repeats of all three
conditions.</li>
<li>Run FastQC on the untrimmed reads only, and note that MultiQC
specifically uses the <code>.zip</code> files for input, not the
<code>.html</code>.</li>
<li>Since multiqc scans for input files, the input names don’t have to
be explicitly mentioned in the <code>shell</code> part.</li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">Show me the solution</h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<pre class="source"><code>rule multiqc:
    output: directory('multiqc_out')
    input:
        salmon =   expand("salmon.{cond}_{rep}", cond=CONDITIONS, rep=REPLICATES),
        kallisto = expand("kallisto.{cond}_{rep}", cond=CONDITIONS, rep=REPLICATES),
        fastqc =   expand("reads.{cond}_{rep}_{end}_fastqc.zip", cond=CONDITIONS, rep=REPLICATES, end=["1","2"]),
    shell:
        "multiqc . -o multiqc_out"</code></pre>
<p>Since the rule has no wildcards, you can run it by giving either the
rule name or the output directory name as a target.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-j1</span> <span class="at">-p</span> multiqc</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="ex">...or</span> equivalently...</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-j1</span> <span class="at">-p</span> multiqc_out</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="fixing-kallisto" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="fixing-kallisto" class="callout-inner">
<h3 class="callout-title">Fixing Kallisto<a class="anchor" aria-label="anchor" href="#fixing-kallisto"></a>
</h3>
<div class="callout-content">
<p>You may notice that MultiQC is not capturing any Kallisto output when
making the reports. The reason for this is given in the <a href="https://multiqc.info/docs/#kallisto" class="external-link">MultiQC manual here</a>:</p>
<blockquote>
<p><em>Note - MultiQC parses the standard out from Kallisto, not any of
its output files (abundance.h5, abundance.tsv, and run_info.json). As
such, you must capture the Kallisto output to a file when running it for
use with MultiQC.</em></p>
</blockquote>
<p>Fix the Snakefile so that Kallisto terminal output is redirected to a
file and can be collected by MultiQC.</p>
<ul>
<li>
<em>Hint 1:</em> The manual above is not quite right - you need to
capture both <strong>stdout and stderr</strong>, so use
<code>&gt;&amp;</code> rather than <code>&gt;</code>, as we did with the
indexing step.</li>
<li>
<em>Hint 2:</em> MultiQC does not mind what you call the file, so
choose your own sensible name.</li>
</ul>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">Show me the solution</h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" aria-labelledby="headingSolution3" data-bs-parent="#accordionSolution3">
<div class="accordion-body">
<pre class="source"><code># Kallisto quantification of one sample, with log capture.
rule kallisto_quant:
    output: directory("kallisto.{sample}")
    input:
        index = "Saccharomyces_cerevisiae.R64-1-1.kallisto_index",
        fq1   = "trimmed/{sample}_1.fq",
        fq2   = "trimmed/{sample}_2.fq",
    shell:
        """mkdir {output}
           kallisto quant -i {input.index} -o {output} {input.fq1} {input.fq2} &gt;&amp; {output}/kallisto_quant.log
        """</code></pre>
<p>There are several perfectly good ways of structuring this, so don’t
worry if your answer is different.</p>
<p>A gotcha with the above version is that the output directory needs to
be created before <em>kallisto quant</em> is run, much like with FastQC.
Remember that Snakemake deletes any existing outputs, including outputs
that are directories, before the job is run, and while Kallisto will
create the directory for you this is too late for the shell to make the
log file and without the <code>mkdir</code> command it will report “No
such file or directory”.</p>
<p>Another option is to make the log file outside of the directory, or
stick with declaring the individual files as outputs, as when we first
made the rule, in which case the directory will be made by Snakemake.
It’s also possible to declare both the directory and a file within that
directory as separate outputs, which is unusual but may be a good
approach here.</p>
</div>
</div>
</div>
</div>
<div id="making-the-multiqc-rule-more-robust" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="making-the-multiqc-rule-more-robust" class="callout-inner">
<h3 class="callout-title">Making the MultiQC rule more robust<a class="anchor" aria-label="anchor" href="#making-the-multiqc-rule-more-robust"></a>
</h3>
<div class="callout-content">
<p>Because MultiQC scans for suitable input rather than taking an
explicit list of files, there is a risk that it picks up unwanted
reports if you leave old files sitting in the directory. To make the
rule fully explicit, one idea is to make a temporary directory and
symlink all the files into it, and then tell MultiQC to look in there.
Amend the <em>multiqc</em> rule so it does this.</p>
<ul>
<li>
<em>Hint 1:</em> When making links, use
<code>ln -snr -t &lt;target_dir&gt; &lt;src&gt;</code> to avoid link
relativity issues.</li>
<li>
<em>Hint 2:</em> If you feel that tweaking some other rules would
make this easier, feel free to do that.</li>
</ul>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4">Show me the solution</h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" aria-labelledby="headingSolution4" data-bs-parent="#accordionSolution4">
<div class="accordion-body">
<p>This solution will work with the version of <em>kallisto_quant</em>
in the solution above.</p>
<pre class="source"><code>rule multiqc:
    output:
        mqc_out = directory('multiqc_out'),
        mqc_in  = directory('multiqc_in'),
    input:
        salmon =   expand("salmon.{cond}_{rep}", cond=CONDITIONS, rep=REPLICATES),
        kallisto = expand("kallisto.{cond}_{rep}", cond=CONDITIONS, rep=REPLICATES),
        fastqc =   expand("reads.{cond}_{rep}_{end}_fastqc.zip", cond=CONDITIONS, rep=REPLICATES, end=["1","2"]),
    shell:
        """mkdir {output.mqc_in}
           ln -snr -t {output.mqc_in} {input}
           multiqc {output.mqc_in} -o {output.mqc_out}
        """</code></pre>
</div>
</div>
</div>
</div>
<p>To see the actual MultiQC report, open the file
<em>multiqc_out/multiqc_report.html</em> in a web browser. You can do
this directly from the command line, assuming you have a default browser
configured.</p>
<p>On Linux environments:</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> xdg-open multiqc_out/multiqc_report.html</span></code></pre>
</div>
<p>For MacOS:</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> open multiqc_out/multiqc_report.html</span></code></pre>
</div>
<p>The report has a few issues, but we’ll not get distracted by the
details of how to configure MultiQC to resolve them.</p>
<p><em>For reference, <a href="files/ep07.Snakefile">this is a
Snakefile</a> incorporating the changes made in this episode. You may
now proceed to any later episode in the lesson using this workflow as a
starting point.</em></p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Once a workflow is complete and working, there will still be room
for refinement</li>
<li>This completes the introduction to the fundamentals of
Snakemake</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-08-configuring"><p>Content from <a href="08-configuring.html">Configuring workflows</a></p>
<hr>
<p>Last updated on 2024-03-11 |
        
        <a href="https://github.com/carpentries-incubator/snakemake-novice-bioinformatics/edit/main/episodes/08-configuring.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do I separate my rules from my configuration?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Add parameters to rules</li>
<li>Use configuration files and command line options to set the
parameters</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<p><em>For reference, <a href="files/ep07.Snakefile">this is the final
Snakefile from episodes 1 to 7</a> you may use to start this
episode.</em></p>
<section id="adding-parameters-params-to-rules"><h2 class="section-heading">Adding parameters (params) to rules<a class="anchor" aria-label="anchor" href="#adding-parameters-params-to-rules"></a>
</h2>
<hr class="half-width">
<p>So far, we’ve written rules with <code>input</code>,
<code>output</code> and <code>shell</code> parts. Another useful section
you can add to a rule is <code>params</code>.</p>
<p>Consider the “trimreads” rule we defined earlier in the course.</p>
<pre class="source"><code>rule trimreads:
    output: "trimmed/{myfile}.fq"
    input:  "reads/{myfile}.fq"
    shell:
        "fastq_quality_trimmer -t 20 -l 100 -o {output} &lt;{input}"</code></pre>
<p>Can you remember what the <code>-t 20</code> and <code>-l 100</code>
parameters do without referring back to the manual? Probably not! Adding
comments in the Snakefile would certainly help, but we can also make
important settings into parameters.</p>
<pre class="source"><code>rule trimreads:
    output: "trimmed/{myfile}.fq"
    input:  "reads/{myfile}.fq"
    params:
        qual_threshold = "20",
        min_length     = "100",
    shell:
        "fastq_quality_trimmer -t {params.qual_threshold} -l {params.min_length} -o {output} &lt;{input}"</code></pre>
<p>Now it is a little clearer what these numbers mean. Use of parameters
doesn’t give you extra functionality but it is good practise put
settings like these into parameters as it makes the whole rule more
readable.</p>
<div id="adding-a-parameter-to-the-salmon_index-rule" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="adding-a-parameter-to-the-salmon_index-rule" class="callout-inner">
<h3 class="callout-title">Adding a parameter to the<code>salmon_index</code>rule<a class="anchor" aria-label="anchor" href="#adding-a-parameter-to-the-salmon_index-rule"></a>
</h3>
<div class="callout-content">
<p>Modify the existing salmon_index rule so that the <code>-k</code>
setting (k-mer length) is a parameter.</p>
<p>Change the length to 29 and re-build the index with the amended
rule.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">Show me the solution</h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<pre class="source"><code>rule salmon_index:
    output:
        idx = directory("{strain}.salmon_index")
    input:
        fasta = "transcriptome/{strain}.cdna.all.fa.gz"
    params:
        kmer_len = "29"
    shell:
        "salmon index -t {input.fasta} -i {output.idx} -k {params.kmer_len}"</code></pre>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-j1</span> <span class="at">-p</span> <span class="at">-f</span> Saccharomyces_cerevisiae.R64-1-1.salmon_index</span></code></pre>
</div>
<p>Notes:</p>
<ul>
<li>You can choose a different parameter name, but it must be a valid
identifier - no spaces or hyphens.</li>
<li>Changing the parameters doesn’t automatically trigger Snakemake to
re-run the rule so you need to use <code>-f</code> (or <code>-R</code>
or <code>-F</code>) to force the job to be re-run (but, as mentioned in
<a href="04-the_dag.html">episode 4</a>, this behaviour is changed in
recent Snakemake versions).</li>
</ul>
</div>
</div>
</div>
</div>
</section><section id="making-snakefiles-configurable"><h2 class="section-heading">Making Snakefiles configurable<a class="anchor" aria-label="anchor" href="#making-snakefiles-configurable"></a>
</h2>
<hr class="half-width">
<p>It’s good practise to break out parameters that you intend to change
into a separate file. That way you can re-run the pipeline on new input
data, or with alternative settings, but you don’t need to edit the
Snakefile itself.</p>
<p>We’ll save the following lines into a file named
<em>config.yaml</em>.</p>
<pre class="source"><code><span><span class="va">salmon_kmer_len</span><span class="op">:</span> <span class="st">"31"</span></span>
<span><span class="va">trimreads_qual_threshold</span><span class="op">:</span> <span class="st">"20"</span></span>
<span><span class="va">trimreads_min_length</span><span class="op">:</span> <span class="st">"100"</span></span></code></pre>
<p>This file is in YAML format. This format allows you to capture
complex data structures but we’ll just use it to store some name+value
pairs. We can then reference these values within the Snakefile via the
<strong>config</strong> object.</p>
<pre class="source"><code>rule trimreads:
    output: "trimmed/{myfile}.fq"
    input:  "reads/{myfile}.fq"
    params:
        qual_threshold = config["trimreads_qual_threshold"],
        min_length     = config.get("trimreads_min_length", "100"),
    shell:
        "fastq_quality_trimmer -t {params.qual_threshold} -l {params.min_length} -o {output} &lt;{input}"</code></pre>
<p>In the above example, the <strong>trimreads_qual_threshold</strong>
value must be supplied in the config, but the
<strong>trimreads_min_length</strong> can be omitted, and then the
default of “100” will be used.</p>
<p>If you are a Python programmer you’ll recognise the syntax here. If
not, then just take note that the first form uses <em>square
brackets</em> and the other uses <code>.get(...)</code> with <em>regular
brackets</em>. Both the config entry name and the default value should
be in quotes.</p>
<div id="using-config-settings-with-params" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="using-config-settings-with-params" class="callout-inner">
<h3 class="callout-title">Using config settings with params<a class="anchor" aria-label="anchor" href="#using-config-settings-with-params"></a>
</h3>
<div class="callout-content">
<p>Strictly speaking, you don’t have to use <em>config</em> in
conjunction with <em>params</em> like this, but it’s normally a good
idea to do so.</p>
</div>
</div>
</div>
<p>The final step is to tell Snakemake about your config file, by
referencing it on the command line:</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">--configfile</span><span class="op">=</span>config.yaml ...</span></code></pre>
</div>
<div id="making-the-parameter-of-a-rule-configurable" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="making-the-parameter-of-a-rule-configurable" class="callout-inner">
<h3 class="callout-title">Making the parameter of a rule
configurable<a class="anchor" aria-label="anchor" href="#making-the-parameter-of-a-rule-configurable"></a>
</h3>
<div class="callout-content">
<p>Fix the <code>salmon_index</code> rule to use
<code>salmon_kmer_len</code> as in the config file sample above. Use a
default of “29” if no config setting is supplied.</p>
<p>Run Snakemake in <em>dry run</em> mode (<code>-n</code>) to check
that this is working as expected.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">Show me the solution</h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<p>Rule is as before, aside from:</p>
<pre class="source"><code><span><span class="va">params</span><span class="op">:</span></span>
<span>    <span class="va">kmer_len</span> <span class="op">=</span> <span class="fu">config.get</span><span class="op">(</span><span class="st">"salmon_kmer_len"</span>, <span class="st">"29"</span><span class="op">)</span></span></code></pre>
<p>If you run Snakemake with the <code>-n</code> and <code>-p</code>
flags and referencing the config file, you should see that the command
being printed has the expected value of <em>31</em>.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-n</span> <span class="at">-p</span> <span class="at">-f</span> <span class="at">--configfile</span><span class="op">=</span>config.yaml Saccharomyces_cerevisiae.R64-1-1.salmon_index</span></code></pre>
</div>
<p><em>Note that if you try to run Snakemake with no config file you
will now get a <strong>KeyError</strong> regarding
<strong>trimreads_qual_threshold</strong>. Even though you are not using
the <strong>trimreads</strong> rule, Snakemake needs a setting for all
mandatory parameters.</em></p>
</div>
</div>
</div>
</div>
<p>Before proceeding, we’ll tweak the Snakefile in a couple of ways:</p>
<ol style="list-style-type: decimal">
<li>Set a default <code>configfile</code> option so we don’t need to
type it on every command line.</li>
<li>Get Snakemake to print out the config whenever it runs.</li>
</ol>
<p>Add the following lines right at the top of the Snakefile.</p>
<pre class="source"><code><span><span class="va">configfile</span><span class="op">:</span> <span class="st">"config.yaml"</span></span>
<span><span class="fu">print</span><span class="op">(</span><span class="st">"Config is: "</span>, <span class="va">config</span><span class="op">)</span></span></code></pre>
<p>Finally, as well as the <code>--configfile</code> option to Snakemake
there is the <code>--config</code> option which sets individual
configuration parameters.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">--configfile</span><span class="op">=</span>config.yaml <span class="at">--config</span> salmon_kmer_len=23 <span class="at">-p</span> <span class="at">-nf</span> Saccharomyces_cerevisiae.R64-1-1.salmon_index/</span></code></pre>
</div>
<p>This is all getting quite complex, so in summary:</p>
<ul>
<li>Snakemake loads the <code>--configfile</code> supplied on the
command line, or else defaults to the one named in the Snakefile, or
else runs with no config file.</li>
<li>Individual <code>--config</code> items on the command line always
take precedence over settings in the config file.</li>
<li>You can set multiple <code>--config</code> values on the command
line and the list ends when there is another parameter, in this case
<code>-p</code>.</li>
<li>Use the <code>config.get("item_name", "default_val")</code> syntax
to supply a default value which takes lowest precedence.</li>
<li>Use <code>config["item_name"]</code> syntax to have a mandatory
configuration option.</li>
</ul>
<div id="making-the-conditions-and-replicates-into-configurable-lists" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="making-the-conditions-and-replicates-into-configurable-lists" class="callout-inner">
<h3 class="callout-title">Making the conditions and replicates into
configurable lists<a class="anchor" aria-label="anchor" href="#making-the-conditions-and-replicates-into-configurable-lists"></a>
</h3>
<div class="callout-content">
<p>Modify the <em>Snakefile</em> and <em>config.yaml</em> so that you
are setting the <em>CONDITIONS</em> and <em>REPLICATES</em> in the
config file. Lists in YAML use the same syntax as Python, with square
brackets and commas, so you can copy the lists you already have. Note
that you’re not expected to modify any rules here.</p>
<p>Re-run the workflow to make a report on <em>just replicates 2 and
3</em>. Check the MultiQC report to see that it really does have just
these replicates in there.</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">Show me the solution</h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<p>In <em>config.yaml</em> add the lines:</p>
<pre class="source"><code>conditions: ["etoh60", "temp33", "ref"]
replicates: ["1", "2", "3"]</code></pre>
<p>In the <em>Snakefile</em> we can reference the <em>config</em> while
setting the global variables. There are no <em>params</em> to add
because these settings are altering the selection of jobs to be added to
the DAG, rather than just the <em>shell</em> commands.</p>
<pre class="source"><code><span><span class="va">CONDITIONS</span> <span class="op">=</span> <span class="va">config</span><span class="op">[</span><span class="st">"conditions"</span><span class="op">]</span></span>
<span><span class="va">REPLICATES</span> <span class="op">=</span> <span class="va">config</span><span class="op">[</span><span class="st">"replicates"</span><span class="op">]</span></span></code></pre>
<p>And for the final part we can either edit the <em>config.yaml</em> or
override on the command line:</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-j1</span> <span class="at">-f</span> <span class="at">--config</span> replicates=[<span class="st">"2"</span>,<span class="st">"3"</span>] <span class="at">-p</span> multiqc_out</span></code></pre>
</div>
<p>Note that we need to re-run the final report, but only this, so only
<code>-f</code> is necessary. If you find that replicate 1 is still in
you report, make sure you are using the final version of the
<em>multiqc</em> rule from the previous episode, that symlinks the
inputs into a <em>multiqc_in</em> directory.</p>
</div>
</div>
</div>
</div>
<p><em>For reference, <a href="files/ep08.Snakefile">this is a
Snakefile</a> incorporating the changes made in this episode.</em></p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Break out significant options into rule parameters</li>
<li>Use a YAML config file to separate your configuration from your
workflow logic</li>
<li>Decide if different config items should be mandatory or else have a
default</li>
<li>Reference the config file in your Snakefile or else on the command
line with <code>--configfile</code>
</li>
<li>Override or add config values using
<code>--config name1=value1 name2=value2</code> and end the list with a
new parameter, eg. <code>-p</code>
</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-09-performance"><p>Content from <a href="09-performance.html">Optimising workflow performance</a></p>
<hr>
<p>Last updated on 2024-03-01 |
        
        <a href="https://github.com/carpentries-incubator/snakemake-novice-bioinformatics/edit/main/episodes/09-performance.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What compute resources are available on my system?</li>
<li>How do I define jobs with more than one thread?</li>
<li>How do I measure the compute resources being used by a
workflow?</li>
<li>How do I run my workflow steps in parallel?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Understand CPU, RAM and I/O bottlenecks</li>
<li>Understand the <em>threads</em> declaration</li>
<li>Use standard Linux tools to look at resource usage</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<p><em>For reference, <a href="files/ep07.Snakefile">this is the final
Snakefile from episodes 1 to 7</a> you may use to start this
episode.</em></p>
<section id="processes-threads-and-processors"><h2 class="section-heading">Processes, threads and processors<a class="anchor" aria-label="anchor" href="#processes-threads-and-processors"></a>
</h2>
<hr class="half-width">
<p>Some definitions:</p>
<ul>
<li>
<strong>Process</strong> - A running program (in our case, each
Snakemake job can be considered one process)</li>
<li>
<strong>Threads</strong> - Each process has one or more threads
which run in parallel</li>
<li>
<strong>Processor</strong> - Your computer has multiple <em>CPU
cores</em> or processors, each of which can run one thread at a
time</li>
</ul>
<p>These definitions are a little simplified, but fine for our needs.
The operating system kernel shares out threads among processors:</p>
<ul>
<li>Having <em>fewer threads</em> than <em>processors</em> means you are
not fully using all your CPU cores</li>
<li>Having <em>more threads</em> than <em>processors</em> means threads
have to “timeslice” on a core which is generally suboptimal</li>
</ul>
<p>If you tell Snakemake how many threads each rule will use, and how
many cores you have available, it will start jobs in parallel to use all
your cores. In the diagram below, five jobs are ready to run and there
are four system cores.</p>
<figure><img src="fig/snake_threads.svg" alt="Representation of a computer with four microchip icons indicating four available cores. To the right are five small green boxes representing Snakemake jobs and labelled as wanting 1, 1, 1, 2 and 8 threads respecively." class="figure mx-auto d-block"></figure></section><section id="listing-the-resources-your-linux-machine"><h2 class="section-heading">Listing the resources your Linux machine<a class="anchor" aria-label="anchor" href="#listing-the-resources-your-linux-machine"></a>
</h2>
<hr class="half-width">
<p>Find out how many CPU cores you have on your machine with the
<code>lscpu</code> command.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> lscpu</span></code></pre>
</div>
<p>Likewise find out the amount of RAM available:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> free <span class="at">-h</span></span></code></pre>
</div>
<p>And finally disk space, on the current partition:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> df <span class="at">-h</span> .</span></code></pre>
</div>
<p>(or <code>df -h</code> without the <code>.</code> to show all
partitions)</p>
</section><section id="parallel-jobs-in-snakemake"><h2 class="section-heading">Parallel jobs in Snakemake<a class="anchor" aria-label="anchor" href="#parallel-jobs-in-snakemake"></a>
</h2>
<hr class="half-width">
<p>You may want to see the relevant part of <a href="https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#threads" class="external-link">the
Snakemake documentation</a>.</p>
<p>We’ll force all the trimming and kallisto steps to re-run by using
the -F flag to Snakemake and time the whole run using the standard
<code>/usr/bin/time -v</code> command. You have to type the command like
this because <code>time</code> is a built-in command in BASH which takes
precedence, so eg:</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> /usr/bin/time <span class="at">-v</span> snakemake <span class="at">-j1</span> <span class="at">-F</span> <span class="at">--</span> kallisto.<span class="dt">{ref</span><span class="op">,</span><span class="dt">temp33</span><span class="op">,</span><span class="dt">etoh60}</span>_<span class="dt">{1</span><span class="op">,</span><span class="dt">2</span><span class="op">,</span><span class="dt">3}</span></span></code></pre>
</div>
<div id="measuring-how-concurrency-affects-execution-time" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="measuring-how-concurrency-affects-execution-time" class="callout-inner">
<h3 class="callout-title">Measuring how concurrency affects execution
time<a class="anchor" aria-label="anchor" href="#measuring-how-concurrency-affects-execution-time"></a>
</h3>
<div class="callout-content">
<p>What is the <em>wallclock time</em> reported by the above command?
We’ll work out the average for the whole class, or if you are working
through the material on your own repeat the measurement three times to
get your own average.</p>
<p>Now change the Snakemake concurrency option to <code>-j2</code> and
then <code>-j4</code>.</p>
<ul>
<li>How does the total execution time change?</li>
<li>What factors do you think limit the power of this setting to reduce
the execution time?</li>
</ul>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">Show me the solution</h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>The time will vary depending on the system configuration but
somewhere around 30 seconds is expected, and this should reduce to
around 25 secs with <code>-j2</code> but higher <code>-j</code> will
produce diminishing returns.</p>
<p>Things that may limit the effectiveness of parallel execution
include:</p>
<ul>
<li>The number of processors in the machine</li>
<li>The number of jobs in the DAG which are independent and can
therefore be run in parallel</li>
<li>The existence of single long-running jobs like
<em>kallisto_index</em>
</li>
<li>The amount of RAM in the machine</li>
<li>The speed at which data can be read from and written to disk</li>
</ul>
</div>
</div>
</div>
</div>
<p>There are <strong>a few gotchas</strong> to bear in mind when using
parallel execution:</p>
<ol style="list-style-type: decimal">
<li>Parallel jobs will use more RAM. If you run out then either your OS
will swap data to disk, or a process will crash</li>
<li>Parallel jobs may trip over each other if they try to write to the
same filename at the same time (this can happen with temporary files,
and in fact is a problem with our current <code>fastqc</code> rule
definition)</li>
<li>The on-screen output from parallel jobs will be jumbled, so save any
output to log files instead</li>
</ol></section><section id="multi-thread-rules-in-snakemake"><h2 class="section-heading">Multi-thread rules in Snakemake<a class="anchor" aria-label="anchor" href="#multi-thread-rules-in-snakemake"></a>
</h2>
<hr class="half-width">
<p>In the diagram at the top, we showed jobs with 2 and 8 threads. These
are defined by adding a <code>threads:</code> part to the rule
definition. We could do this for the <em>kallisto_quant</em> rule:</p>
<pre class="source"><code>rule kallisto_quant:
    output:
        outdir = directory("kallisto.{sample}"),
    input:
        index = "Saccharomyces_cerevisiae.R64-1-1.kallisto_index",
        fq1   = "trimmed/{sample}_1.fq",
        fq2   = "trimmed/{sample}_2.fq",
    threads: 4
    shell:
        "kallisto quant -t {threads} -i {input.index} -o {output.outdir} {input.fq1} {input.fq2}"</code></pre>
<p>You should explicitly use <code>threads: 4</code> rather than
<code>params: threads = "4"</code> because Snakemake considers the
number of threads when scheduling jobs. Also, if the number of threads
requested for a rule is less than the number of available processors
then Snakemake will use the lower number.</p>
<p>We also added <code>-t {threads}</code> to the shell command. This
only works for programs which allow you to specify the number of threads
as a command-line option, but this applies to a lot of different
bioinformatics tools.</p>
<div id="getting-other-programs-to-run-with-multiple-threads" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="getting-other-programs-to-run-with-multiple-threads" class="callout-inner">
<h3 class="callout-title">Getting other programs to run with multiple
threads<a class="anchor" aria-label="anchor" href="#getting-other-programs-to-run-with-multiple-threads"></a>
</h3>
<div class="callout-content">
<p>Find out how to set the number of threads for our
<em>salmon_quant</em> and <em>fastqc</em> jobs. Which of the options
below would need to be added to the shell command in each case?</p>
<ol style="list-style-type: decimal">
<li><code>-t {threads}</code></li>
<li><code>-p {threads}</code></li>
<li><code>-num_threads {threads}</code></li>
<li>multi-threaded mode is not supported</li>
</ol>
<p><em>Hint: use <code>salmon quant --help-alignment</code> and
<code>fastqc --help</code>, or search the online documentation.</em></p>
<p>Make the corresponding changes to the Snakefile.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">Show me the solution</h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<p>For <em>salmon_quant</em>, <code>-p {threads}</code> or equivalently
<code>--threads {threads}</code> will work.</p>
<p>For <em>fastqc</em>, it may look like <code>-t {threads}</code> is
good but this only sets “the number of files which can be processed
simultaneously”, and the rule we have only processes a single file per
job. So in fact the answer is that, for our purposes, multi-threading is
unsupported.</p>
</div>
</div>
</div>
</div>
<div id="fine-grained-profiling" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="fine-grained-profiling" class="callout-inner">
<h3 class="callout-title">Fine-grained profiling<a class="anchor" aria-label="anchor" href="#fine-grained-profiling"></a>
</h3>
<div class="callout-content">
<p>Rather than timing the entire workflow, we can ask Snakemake to
benchmark an individual rule.</p>
<p>For example, to benchmark the <code>kallisto_quant</code> step we
could add this to the rule definition:</p>
<pre class="source"><code>rule kallisto_quant:
    benchmark:
        "benchmarks/kallisto_quant.{sample}.txt"
    ...</code></pre>
<p>The dataset here is so small that the numbers are tiny, but for real
data this can be very useful as it shows time, memory usage and IO load
for all jobs.</p>
</div>
</div>
</div>
</section><section id="running-jobs-on-a-cluster"><h2 class="section-heading">Running jobs on a cluster<a class="anchor" aria-label="anchor" href="#running-jobs-on-a-cluster"></a>
</h2>
<hr class="half-width">
<p>Learning about clusters is beyond the scope of this course, but for
modern bioinformatics they are an essential tool because many analysis
jobs would take too long on a single computer. Learning to run jobs on
clusters normally means writing batch scripts and re-organising your
code to be cluster-aware. But if your workflow is written in Snakemake,
it will run on a cluster will little to no modification. Snakemake turns
the individual jobs into cluster jobs, then submits and monitors them
for you.</p>
<ul>
<li><a href="https://snakemake.readthedocs.io/en/stable/executing/cluster.html" class="external-link">The
Snakemake manual explains how to set this up</a></li>
<li><a href="files/snakemake_on_eddie.pdf">We have some specific
suggestions for Eddie, the University of Edinburgh cluster</a></li>
</ul>
<figure><img src="fig/Multiple_Server_.jpg" alt="A photo of some high performance computer hardware racked in five cabinets in a server room. Each cabinet is about 2.2 metres high and 0.8m wide. The doors of the cabinets are open to show the systems inside. Orange and yellow cabling is prominent, connecting ports within the second and third racks." class="figure mx-auto d-block"></figure><div id="cluster-demo" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="cluster-demo" class="callout-inner">
<h3 class="callout-title">Cluster demo<a class="anchor" aria-label="anchor" href="#cluster-demo"></a>
</h3>
<div class="callout-content">
<p>A this point in the course there may be a cluster demo…</p>
</div>
</div>
</div>
<p>{% comment %} Photo credit: Cskiran Sourced from Wikimedia Commons
CC-BY-SA-4.0 {% endcomment %}</p>
<p><em>For reference, <a href="files/ep09.Snakefile">this is a
Snakefile</a> incorporating the changes made in this episode.</em></p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>To make your workflow run as fast as possible, try to match the
number of threads to the number of cores you have</li>
<li>You also need to consider RAM, disk, and network bottlenecks</li>
<li>Profile your jobs to see what is taking most resources</li>
<li>Snakemake is great for running workflows on compute clusters</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-10-conda_integration"><p>Content from <a href="10-conda_integration.html">Conda integration</a></p>
<hr>
<p>Last updated on 2024-03-11 |
        
        <a href="https://github.com/carpentries-incubator/snakemake-novice-bioinformatics/edit/main/episodes/10-conda_integration.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do I install new packages with Conda?</li>
<li>How do I get Snakemake to manage software dependencies?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Understand how Snakemake works with Conda and Bioconda</li>
<li>Modify a package in the workflow with Conda integration</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<p><em>For reference, <a href="files/ep07.Snakefile">this is the final
Snakefile from episodes 1 to 7</a> you may use to start this
episode.</em></p>
<section id="the-basics-of-conda"><h2 class="section-heading">The basics of Conda<a class="anchor" aria-label="anchor" href="#the-basics-of-conda"></a>
</h2>
<hr class="half-width">
<p>You may already have some familiarity with the basics of <a href="https://conda.io" class="external-link">Conda</a> and <a href="https://bioconda.github.io/" class="external-link">Bioconda</a>.</p>
<p>Some key terms:</p>
<ul>
<li>
<strong>Conda</strong> is a system for installing software
<strong>packages</strong> into self-contained directories called
<strong>environments</strong>.</li>
<li>
<strong>Conda</strong> finds new packages in on-line repositories
which are known as <strong>channels</strong>.</li>
<li>The <strong>Bioconda project</strong> maintains a channel with many
open source bioinformatics packages available.</li>
<li>Old versions of these packages are available, as well as the latest
builds.</li>
<li>Any environment may have multiple packages, but to install two
versions of the same package, or two packages which conflict, you need
to put them in <strong>separate environments</strong>.</li>
<li>You can switch between environments using the <strong>conda
activate</strong> command.</li>
<li>An environment may be <strong>exported</strong>, which simply means
getting Conda to print all the packages in that environment, as well as
the channels where those packages are available, in a YAML format.</li>
</ul>
<p>Assuming you are running throught this course in the <a href="index.html#software-installation">recommended setup</a>, Conda is
already set up on the systems we are using.</p>
<p>Some key conda commands:</p>
<ul>
<li>Make a new environment and give it a name
<ul>
<li><code>$ conda create -n my-environment</code></li>
</ul>
</li>
<li>List the environments currently known to conda
<ul>
<li><code>$ conda env list</code></li>
</ul>
</li>
<li>Switch to a named environment in the current shell session
<ul>
<li><code>$ conda activate my-environment</code></li>
</ul>
</li>
<li>Configure Conda to enable the Bioconda channel (the new settings
will apply to all environments)
<ul>
<li><code>$ conda config --add channels bioconda</code></li>
<li><code>$ conda config --add channels conda-forge</code></li>
</ul>
</li>
<li>Install the latest version of a single package into the current
environment
<ul>
<li><code>$ conda install snakemake</code></li>
</ul>
</li>
<li>Install a specific version of a single package into the current
environment
<ul>
<li><code>$ conda install snakemake=5.14.0</code></li>
</ul>
</li>
<li>Save a list of all packages in one environment in YAML format, then
recreate an identical environment from that information.
<ul>
<li><code>$ conda env export -n my-environment &gt; my-environment.yaml</code></li>
<li><code>$ conda env create -n cloned-environment -f my-environment.yaml</code></li>
</ul>
</li>
</ul>
<div id="channel-configuration-and-conda-forge" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="channel-configuration-and-conda-forge" class="callout-inner">
<h3 class="callout-title">Channel configuration and conda-forge<a class="anchor" aria-label="anchor" href="#channel-configuration-and-conda-forge"></a>
</h3>
<div class="callout-content">
<p>For advanced usage, there are many ways you might want to configure
your Conda channels. It’s even possible to have specific channel
settings for each environment. For our purposes, we just want to have
the <em>bioconda</em> channel working, and as noted on <a href="https://bioconda.github.io/user/install.html#set-up-channels" class="external-link">the
Bioconda website</a> this involves a dependency a second channel named
<em>conda-forge</em>, which provides some supporting tools and
libraries.</p>
<p>To be sure all is well, check your channel settings:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> cat ~/.condarc</span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="va">channels</span><span class="op">:</span></span>
<span>  <span class="op">-</span> <span class="va">conda</span><span class="op">-</span><span class="va">forge</span></span>
<span>  <span class="op">-</span> <span class="va">bioconda</span></span>
<span><span class="va">solver</span><span class="op">:</span> <span class="va">libmamba</span></span>
<span><span class="va">channel_priority</span><span class="op">:</span> <span class="va">strict</span></span></code></pre>
</div>
<p>You may have other channels listed after “bioconda” but you should
have these two at the top, and the other settings. Once this
configuration is right you won’t need to do anything else regarding the
channel configuration in this course.</p>
</div>
</div>
</div>
<div id="queryting-and-installing-conda-packages" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="queryting-and-installing-conda-packages" class="callout-inner">
<h3 class="callout-title">Queryting and installing Conda packages<a class="anchor" aria-label="anchor" href="#queryting-and-installing-conda-packages"></a>
</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li><p>Find out what version of <em>fastx_toolkit</em> is installed in
the current conda environment.</p></li>
<li><p>Create a new conda environment named <em>new-env</em> and install
the <em>cutadapt</em> package from Bioconda into it.</p></li>
</ol>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">Show me the solution</h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<ol style="list-style-type: decimal">
<li>There are several ways of doing this, but one using the commands
above is:</li>
</ol>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> conda env export <span class="kw">|</span> <span class="fu">grep</span> fastx</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a> <span class="ex">-</span> fastx_toolkit=0.0.14=0</span></code></pre>
</div>
<ol start="2" style="list-style-type: decimal">
<li>Conda should already be configured to install Bioconda packages (see
the callout above) so we can do this:</li>
</ol>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> conda create <span class="at">-n</span> new-env</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> conda activate new-env</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> conda install cutadapt</span></code></pre>
</div>
<p>It’s also possible to install packages at the same time as creating
the environment, though this wasn’t shown in the examples above.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> conda create <span class="at">-n</span> new-env cutadapt</span></code></pre>
</div>
<p>You’ll still need to <em>activate</em> the new environment in order
to run <em>cutadapt</em>.</p>
</div>
</div>
</div>
</div>
</section><section id="using-conda-with-snakemake"><h2 class="section-heading">Using Conda with Snakemake<a class="anchor" aria-label="anchor" href="#using-conda-with-snakemake"></a>
</h2>
<hr class="half-width">
<p>Up to now, the <code>shell</code> commands you have been running via
Snakemake have called programs in your default <em>PATH</em>. The
commands may have been installed with conda, or with the system package
manager, or installed manually. If you move your Snakefile to another
machine, or share it with a colleague, you will need to make sure all
the dependencies are installed. If the versions of the packages are not
the same on the two systems, you may discover that the workflow breaks,
or produces different results.</p>
<p>Once you are familiar with Conda, you may think to install the
dependencies for your workflow into a Conda environment, then
<code>conda env export</code> that into a YAML file, which you can use
to quickly set up the same environment on any other machine.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> conda env export <span class="at">-n</span> new-env <span class="op">&gt;</span> new-env.yaml</span></code></pre>
</div>
<p>Snakemake takes this one step further by integrating directly with
Conda. This brings some nice features…</p>
<p>Firstly it allows you to specify different environments to use for
different rules. This is really useful if different rules need mutually
incompatible packages.</p>
<p>To configure this, add a <code>conda</code> declaration to a
rule:</p>
<pre class="source"><code>rule a_conda_rule:
    conda:  "new-env.yaml"
    shell:
        "which cutadapt"</code></pre>
<p>Note that the declaration refers to the exported YAML file, not any
existing environment. Snakemake will create the environment for you. We
can now run the above rule, even though it doesn’t produce any outputs.
We need to add the <code>--use-conda</code> option when running
Snakemake.</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>$ snakemake -j1 --use-conda a_conda_rule
Building DAG of jobs...
Creating conda environment new-env.yaml...
Downloading and installing remote packages.
Environment for new-env.yaml created (location: .snakemake/conda/d7df5e24)
Using shell: /bin/bash
Provided cores: 1 (use --cores to define parallelism)
Rules claiming more threads will be scaled down.
Job counts:
    count   jobs
    1       a_conda_rule
    1
Select jobs to execute...

[Tue Sep 21 16:57:59 2021]
rule a_conda_rule:
    jobid: 0

Activating conda environment: /home/zenmaster/carpentries/nextflow_rnaseq_training_dataset/.snakemake/conda/d7df5e24
/home/zenmaster/carpentries/nextflow_rnaseq_training_dataset/.snakemake/conda/d7df5e24/bin/cutadapt
[Tue Sep 21 16:58:00 2021]
Finished job 0.
1 of 1 steps (100%) done
Complete log: /home/zenmaster/carpentries/nextflow_rnaseq_training_dataset/.snakemake/log/2021-09-21T165713.679409.snakemake.log</code></pre>
</div>
<p>This takes some time, because Snakemake is (as the log says) creating
the new environment and installing the packages. But if we run it a
second time it’s much quicker. This brings us to a second feature of
Snakemake+Conda integration: as long as the YAML file is unchanged,
Snakemake will re-use the same environment, but if the file is edited or
replaced Snakemake will detect the change and make a new environment.
This happens because the environment that Snakemake made is stored under
<code>./.snakemake/conda/d7df5e24</code>, and the last part of this name
is a hash of the contents of the <code>new-env.yaml</code> file.</p>
<p>Also, it may seem very inefficient to create the environment twice,
but Conda employs file de-duplication so you don’t actually end up with
two copies of all the software.</p>
<p>We’ll do something useful with <em>cutadapt</em> in the next
episode.</p>
<div id="using-an-older-version-of-salmon" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="using-an-older-version-of-salmon" class="callout-inner">
<h3 class="callout-title">Using an older version of Salmon<a class="anchor" aria-label="anchor" href="#using-an-older-version-of-salmon"></a>
</h3>
<div class="callout-content">
<p>Going back to our RNA-Seq workflow, imagine we want to try running
the analysis with an older version of Salmon, to see if the results are
different. We’ll use <strong>Salmon 1.2.1</strong> which is available in
Bioconda.</p>
<p>This particular package happens to need a dependency,
<code>tbb=2020.2</code>, which is not installed by default but we can
explicitly install it into the environment with Salmon. Without this
Salmon 1.2.1 will crash out with the message <em>error while loading
shared libraries</em>.</p>
<p>We don’t want to mess with the version of Salmon that’s currently
installed, or change any parts of the workflow other than adding
directives to <em>salmon_index</em> and <em>salmon_quant</em>. Define a
new Conda environment that includes the packages
<code>salmon=1.2.1</code> and <code>tbb=2020.2</code> and then use this
for the two rules by adding appropriate <code>conda:</code> directives.
Then run your amended workflow.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">Show me the solution</h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<p>The first thing that we need is an appropriate YAML file. It’s
possible to write one from scratch in a text editor, but for this answer
we’ll follow the process shown above, that is to <em>conda create</em>
an environment, install the required packages and then <em>export</em>
it.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> conda create <span class="at">-n</span> salmon-1.2.1</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> conda activate salmon-1.2.1</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> conda install salmon=1.2.1 tbb=2020.2</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> conda env export <span class="at">-n</span> salmon-1.2.1 <span class="op">&gt;</span> salmon-1.2.1.yaml</span></code></pre>
</div>
<p>As an aside, having done this, we could delete the environment as we
just need the exported YAML file.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> conda deactivate</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> conda env remove <span class="at">-n</span> salmon-1.2.1</span></code></pre>
</div>
<p>Next add the same <em>conda</em> declaration to both the
<em>salmon_index</em> and <em>salmon_quant</em> rules.</p>
<pre class="source"><code><span><span class="va">conda</span><span class="op">:</span> <span class="st">"salmon-1.2.1.yaml"</span></span></code></pre>
<p>And when running the workflow, we need to give the
<code>--use-conda</code> option as well as telling Snakemake that these
two rules have changed and must be re-run.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-j1</span> <span class="at">-Rsalmon_index</span> <span class="at">-Rsalmon_quant</span> <span class="at">--use-conda</span> ...</span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="note-on-the-environment-files" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="note-on-the-environment-files" class="callout-inner">
<h3 class="callout-title">Note on the environment files<a class="anchor" aria-label="anchor" href="#note-on-the-environment-files"></a>
</h3>
<div class="callout-content">
<p>If you look into a YAML file created with
<code>conda env export</code> you will see that Conda lists every single
package dependency and the list is quite large. You may prefer to write
your own YAML file where you can be as precise as you like about which
packages are needed and which version or versions are acceptable. Conda
will fill in the blanks to ensure all dependencies are met.</p>
<p>A file for an environment with cutadapt could be as simple as
this.</p>
<pre class="source"><code><span><span class="va">name</span><span class="op">:</span> <span class="va">cutadapt</span><span class="op">-</span><span class="va">env</span></span>
<span><span class="va">channels</span><span class="op">:</span></span>
<span>  <span class="op">-</span> <span class="va">conda</span><span class="op">-</span><span class="va">forge</span></span>
<span>  <span class="op">-</span> <span class="va">bioconda</span></span>
<span><span class="va">dependencies</span><span class="op">:</span></span>
<span>  <span class="op">-</span> <span class="va">cutadapt</span><span class="op">=</span><span class="fl">3.4</span></span></code></pre>
<p>In this case, Conda will install version 3.4 of Cutadapt, but will
meet the required dependencies by installing the newest packages found
in the channels, so the exact environment will not be the same each
time. This may make the workflow more compatible if, for example, you
try to switch from Linux to a Mac, but it may also cause problems if the
newer packages somehow don’t work properly with Cutadapt 3.4, as happens
with the <code>tbb</code> package above. Sadly, while Conda is a great
tool for installing and managing software it does have quirks and
shortcomings, and software setup continues to be a perennial headache
for bioinformaticians.</p>
</div>
</div>
</div>
<p><em>For reference, <a href="files/ep10.Snakefile">this is a
Snakefile</a> incorporating the changes made in this episode.</em></p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Conda is a system for managing software packages in self-contained
environments</li>
<li>Snakemake rules may be associated with specific Conda
environments</li>
<li>When run with the <code>--use-conda</code> option, Snakemake will
set these up for you</li>
<li>Conda gives you fine control over software versions, without
modifying globally-installed packages</li>
<li>Workflows made this way are super-portable, because Conda handles
installing the correct versions of everything</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-11-assembly_challenge"><p>Content from <a href="11-assembly_challenge.html">Constructing a whole new workflow</a></p>
<hr>
<p>Last updated on 2024-03-01 |
        
        <a href="https://github.com/carpentries-incubator/snakemake-novice-bioinformatics/edit/main/episodes/11-assembly_challenge.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do I approach making a new workflow from scratch?</li>
<li>How do I understand and debug the errors I see?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Apply what you have learned so far in a new example</li>
<li>Understand how to diagnose errors in a workflow</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="assembling-the-sequences"><h2 class="section-heading">Assembling the sequences<a class="anchor" aria-label="anchor" href="#assembling-the-sequences"></a>
</h2>
<hr class="half-width">
<p>This episode introduces a longer challenge to build an assembly
workflow from scratch, using the concepts we’ve learned through the
course.</p>
<p>A de-novo assembly of the RNA sequences in the demo dataset does not
produce a biologically meaningful assembly, but nevertheless attempting
to assemble them yields a nice example workflow, so we will go ahead and
try. Look at <a href="files/assembly_script.sh">the following shell
script</a>:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!/bin/bash</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co"># This script needs the following Bioconda packages:</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="co">#   cutadapt</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="co">#   velvet</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="co">#   bbmap</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="co">#  Use cutadapt to remove adapters. Note that cutadapt works on the paired sequences</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> cutadapt</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="ex">cutadapt</span> <span class="at">-a</span> AGATCGGAAGAGC <span class="at">-A</span> AGATCGGAAGAGC <span class="at">-o</span> cutadapt/ref_1_1.fq <span class="at">-p</span> cutadapt/ref_1_2.fq reads/ref_1_1.fq reads/ref_1_2.fq</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="ex">cutadapt</span> <span class="at">-a</span> AGATCGGAAGAGC <span class="at">-A</span> AGATCGGAAGAGC <span class="at">-o</span> cutadapt/ref_2_1.fq <span class="at">-p</span> cutadapt/ref_2_2.fq reads/ref_2_1.fq reads/ref_2_2.fq</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="ex">cutadapt</span> <span class="at">-a</span> AGATCGGAAGAGC <span class="at">-A</span> AGATCGGAAGAGC <span class="at">-o</span> cutadapt/ref_3_1.fq <span class="at">-p</span> cutadapt/ref_3_2.fq reads/ref_3_1.fq reads/ref_3_2.fq</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Combine all the samples. We still want to keep the read pairs separate.</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> concatenate</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> cutadapt/ref_1_1.fq cutadapt/ref_2_1.fq cutadapt/ref_3_1.fq <span class="op">&gt;</span> concatenate/ref_1.fq</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span> cutadapt/ref_1_2.fq cutadapt/ref_2_2.fq cutadapt/ref_3_2.fq <span class="op">&gt;</span> concatenate/ref_2.fq</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a><span class="co"># Assemble the concatenated files with Velvet. This is a 2-step process: velveth then velvetg</span></span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a><span class="va">kmer_len</span><span class="op">=</span>21</span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a><span class="ex">velveth</span> velvet_tmp_ref_<span class="va">${kmer_len}</span> <span class="va">${kmer_len}</span> <span class="at">-shortPaired</span> <span class="at">-fastq</span> <span class="at">-separate</span> concatenate/ref_1.fq concatenate/ref_2.fq</span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a><span class="ex">velvetg</span> velvet_tmp_ref_<span class="va">${kmer_len}</span></span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a><span class="fu">mv</span> velvet_tmp_ref_<span class="va">${kmer_len}</span>/contigs.fa contigs.fa</span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a><span class="co"># Find the longest contig using the BBMap stats script</span></span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a><span class="ex">stats.sh</span> contigs.fa <span class="kw">|</span> <span class="fu">grep</span> <span class="st">'Max contig length:'</span> <span class="op">&gt;</span> max_contig.txt</span></code></pre>
</div>
<p>The following figure presents the steps involved in the above script,
when run on the <code>ref</code> samples. You will see it already looks
a little like a Snakemake DAG.</p>
<figure><img src="fig/assembly_flow.svg" alt='A box-and-arrow representation of the cutadapt, concatenation, and assembly steps in the above script, with the names of the six files from the three ref samples in pairs at the top. Arrows come down from each pair of files into a corresponding cutadapt box, then under this the arrows cross as they go from the three cutadapt boxes to the two boxes labelled "concatenate". Under this is a single box labelled "assembly (velvet)" and a final box at the bottom labelled "find longest contig".' class="figure mx-auto d-block"></figure><div id="running-the-bash-script" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="running-the-bash-script" class="callout-inner">
<h3 class="callout-title">Running the Bash script<a class="anchor" aria-label="anchor" href="#running-the-bash-script"></a>
</h3>
<div class="callout-content">
<p>Run the script as-is. In order to make it work, you should make a new
conda environment named “assembly-env” with the three packages
installed.</p>
<p>How long is the longest contig found in the assembly that is
produced?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">Show me the solution</h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>The commands to build and activate the Conda env can be found in the
“Conda Integration” chapter.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> conda create <span class="at">-n</span> assembly-env <span class="at">--channel</span> bioconda <span class="at">--channel</span> conda-forge cutadapt velvet bbmap</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> conda activate assembly-env</span></code></pre>
</div>
<p>Now, after saving the script above into a text file, make it
executable and run it.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> chmod +x assembly_script.sh</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> ./assembly_script.sh</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> cat max_contig.txt</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="ex">Max</span> contig length:        655</span></code></pre>
</div>
<p>Don’t worry if you get a slightly different number. We’ve not
specified the exact version of the assembler so the final assembly may
be slightly different if the software has been updated.</p>
</div>
</div>
</div>
</div>
</section><section id="general-tips-for-designing-a-snakemake-workflow"><h2 class="section-heading">General tips for designing a Snakemake workflow<a class="anchor" aria-label="anchor" href="#general-tips-for-designing-a-snakemake-workflow"></a>
</h2>
<hr class="half-width">
<div class="section level3">
<h3 id="design-phase">Design phase<a class="anchor" aria-label="anchor" href="#design-phase"></a>
</h3>
<ul>
<li>Start out with pen and paper. Sketch how many rules you will have
and how you expect the DAG to look.</li>
<li>List any parameters or settings that might need to be adjusted.</li>
<li>Work out which rules (if any) aggregate or split inputs (these may
need input functions).</li>
<li>Get the <code>input</code> and <code>output</code> parts of your
Snakemake rules right before worrying about the <code>shell</code>
sections. Remember that Snakemake builds the DAG before actually running
the <code>shell</code> commands so you should <code>--dryrun</code> the
workflow before running it for real, or even before you put all the
<code>shell</code> commands in.</li>
<li>Snakemake starts matching rules from the final target file to the
input files, so you might prefer to work on your rules in this same
order, starting by writing the last rule first.</li>
<li>Choose the names for your <em>input</em>s, <em>output</em>s and
<em>{wildcards}</em> carefully to make your Snakefile as readable as
possible.</li>
</ul>
</div>
<div class="section level3">
<h3 id="debugging-a-snakemake-workflow">Debugging a Snakemake workflow<a class="anchor" aria-label="anchor" href="#debugging-a-snakemake-workflow"></a>
</h3>
<p>As with all programming jobs, you will inevitably see bugs and errors
the first time you try to run a new Snakefile, because it’s very easy to
make typos. Sometimes Snakemake will give you a very useful and precise
error report, but other times less so. When trying to diagnose a
problem, remember that you need to establish which phase of execution
failed, then look at the most common error causes for that phase.</p>
<ul>
<li>
<strong>Parsing phase failures <em>(before Snakemake says ‘Building
DAG of jobs’)</em>:</strong>
<ul>
<li>Syntax errors
<ul>
<li>Missing commas or colons</li>
<li>Unbalanced quotes or brackets</li>
<li>Wrong indentation</li>
<li>…etc</li>
</ul>
</li>
<li>Failure to evaluate expressions
<ul>
<li>Any Python logic or you added outside of rules</li>
<li>Problems in functions (eg. <code>expand()</code>) in rule inputs or
outputs</li>
</ul>
</li>
<li>Other problems with rule definitions
<ul>
<li>Invalid rule names or declarations</li>
<li>Invalid wildcard names</li>
<li>Mismatched wildcards</li>
</ul>
</li>
</ul>
</li>
<li>
<strong>DAG building failures <em>(before Snakemake tries to run any
jobs)</em>:</strong>
<ul>
<li>Failure to determine target output files</li>
<li>Multiple rules can make the same output</li>
<li>No rule to make a target file
<ul>
<li>Sometimes it is not clear why Snakemake wants to make that
particular file</li>
</ul>
</li>
<li>Circular dependency found (violating the “acyclic” property of a
DAG).</li>
<li>An output file is write-protected</li>
</ul>
</li>
<li>
<strong>DAG running failures <em>(–dry-run works as expected, but a
real run fails)</em>:</strong>
<ul>
<li>If a job fails, Snakemake will report an error, delete all output
files for that job, and stop.</li>
<li>Steps can fail if:
<ul>
<li>A placeholder in {curlies} is not a valid variable.</li>
<li>Any shell commands return non-zero status.</li>
<li>You reference any <em>$shell_variable</em> which has not been
set.</li>
<li>Any expected output file is missing after the commands have
run.</li>
</ul>
</li>
</ul>
</li>
</ul>
<p>If you get syntax errors, look out for unexpected or missing text
colouring in GEdit which might hint at the problem. Punctuation errors
are very common: brackets, quotes, colons, line indentation and commas
all need to be just right. If you can’t see the problem, ask a
demonstrator for help. Often just having a second pair of eyes on your
code is all that is needed to spot a problem.</p>
<div id="challenge---building-a-full-snakemake-workflow" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge---building-a-full-snakemake-workflow" class="callout-inner">
<h3 class="callout-title">Challenge - building a full Snakemake
workflow<a class="anchor" aria-label="anchor" href="#challenge---building-a-full-snakemake-workflow"></a>
</h3>
<div class="callout-content">
<p>Design a Snakemake workflow based upon the above script which will
assemble all reads from each of the three conditions (ref, etoh60 and
temp33) and will do so at four different kmer lengths (19, 21, 25 and
27), so twelve assemblies in total. Use your workflow to discover the
length of the longest contig in each of the twelve assemblies.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">Show me the solution</h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<p>A sample solution to this exercise <a href="files/assembly_with_conda.Snakefile">is available here</a>, along
with a suitable <a href="files/assembly_conda_env.yaml">Conda
environment spec</a>. However, there is no single “correct” answer, so
don’t worry if your approach looks different. Hopefully we will have
time in the course to look through and compare some different
answers.</p>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>By now, you are ready to start using Snakemake for your own workflow
tasks</li>
<li>You may wish to replace an existing script with a Snakemake
workflow</li>
<li>Don’t be disheartened by errors, which are normal; use a systematic
approach to diagnose the problem</li>
</ul>
</div>
</div>
</div>
</div>
</section></section><section id="aio-12-cleaning_up"><p>Content from <a href="12-cleaning_up.html">Cleaning up</a></p>
<hr>
<p>Last updated on 2024-03-11 |
        
        <a href="https://github.com/carpentries-incubator/snakemake-novice-bioinformatics/edit/main/episodes/12-cleaning_up.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do I save disk space by removing temporary files?</li>
<li>How do I protect important outputs from deletion?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Understand the function of <em>temporary</em> outputs.</li>
<li>Learn about running in <em>--touch</em> mode</li>
<li>Learn about <em>shadow</em> mode rules</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<p><em>For reference, <a href="files/ep07.Snakefile">this is the final
Snakefile from episodes 1 to 7</a> you may use to start this
episode.</em></p>
<section id="temporary-files-in-snakemake"><h2 class="section-heading">Temporary files in Snakemake<a class="anchor" aria-label="anchor" href="#temporary-files-in-snakemake"></a>
</h2>
<hr class="half-width">
<p>Analyses in bioinformatics inevitably generate a lot of files. Not
all of these need to be kept, if they are intermediate files produced
during the workflow. Even if the files are not completely redundant,
knowing that you can regenerate them again with a single Snakemake
command means you may well want to clean them up. This way you save disk
space and reduce clutter.</p>
<p>Remember that Snakemake only re-generates intermediate files if it
needs them to make a target, so simply deleting intermediate files
manually works fine. Assuming you already made the <em>multiqc</em>
report, and have not edited the rules in the Snakefile since:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>$ rm trimmed/ref*.fq
$ snakemake -j1 -p multiqc_out
...
Nothing to be done.</code></pre>
</div>
<p>To get Snakemake to clean up files for you, mark them with the
<code>temporary()</code> function. Much like the
<code>directory()</code> function, this is applied only on the outputs
of rules. Any file marked as <em>temporary</em> will be removed by
Snakemake as soon as it is no longer needed.</p>
<p>To provide a better example, lets say we decide to compress the
trimmed reads with <em>gzip</em>. It’s very common to store FASTQ files
in this compressed format, and most software will read the gzipped files
directly, so there is no need to keep the uncompressed files. Add a new
rule like so:</p>
<pre class="source"><code>rule gzip_fastq:
    output: "{afile}.fq.gz"
    input:  "{afile}.fq"
    shell:
        "gzip -nc {input} &gt; {output}"</code></pre>
<p>Note that this will compress any <code>.fq</code> file in any
subdirectory, because Snakemake wildcards can match full paths. Now
modify just the <em>salmon_quant</em> rule to work on the compressed
files.</p>
<pre class="source"><code>rule salmon_quant:
    output: directory("salmon.{sample}")
    input:
        index = "Saccharomyces_cerevisiae.R64-1-1.salmon_index",
        fq1   = "trimmed/{sample}_1.fq.gz",
        fq2   = "trimmed/{sample}_2.fq.gz",
    ...</code></pre>
<p>Finally, declare the output of the <em>trimreads</em> rule to be
<em>temporary</em>.</p>
<pre class="source"><code>rule trimreads:
    output: temporary("trimmed/{sample}.fq")</code></pre>
<p>And now re-run the workflow. Since we modified the <em>trimreads</em>
rule, Snakemake should see that it needs to be rerun:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>$ snakemake -j1 -p multiqc_out
...
$ ls trimmed/
etoh60_1_1.fq.gz  ref_1_1.fq.gz  temp33_1_1.fq.gz
etoh60_1_2.fq.gz  ref_1_2.fq.gz  temp33_1_2.fq.gz
etoh60_2_1.fq.gz  ref_2_1.fq.gz  temp33_2_1.fq.gz
etoh60_2_2.fq.gz  ref_2_2.fq.gz  temp33_2_2.fq.gz
etoh60_3_1.fq.gz  ref_3_1.fq.gz  temp33_3_1.fq.gz
etoh60_3_2.fq.gz  ref_3_2.fq.gz  temp33_3_2.fq.gz</code></pre>
</div>
<p>Snakemake kept the uncompressed trimmed reads long enough to run
<em>kallisto_quant</em> on all the samples, then removed them leaving
only the gzipped versions.</p>
<div id="working-with-gzipped-files" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="working-with-gzipped-files" class="callout-inner">
<h3 class="callout-title">Working with gzipped files<a class="anchor" aria-label="anchor" href="#working-with-gzipped-files"></a>
</h3>
<div class="callout-content">
<p>Like Salmon, Kallisto can read compressed <code>.fq.gz</code> files
directly. Amend the <em>kallisto_quant</em> rule to use gzipped input
files. Show that the new rule can be re-run on all samples without
triggering the <em>trimreads</em> or <em>gzip_fastq</em> rules.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">Show me the solution</h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>In the first instance, this involves changing two input lines in the
<em>kallisto_quant</em> rule definition.</p>
<pre class="source"><code>    fq1   = "trimmed/{sample}_1.fq.gz",
    fq2   = "trimmed/{sample}_2.fq.gz",</code></pre>
<p>We can check the result by trying a <em>dry run</em> (<code>-n</code>
option):</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-n</span> <span class="at">-p</span> multiqc_out</span></code></pre>
</div>
<p>You should see that Snakemake wants to run the
<em>kallisto_quant</em> and <em>multiqc</em> steps but no others.</p>
</div>
</div>
</div>
</div>
<div id="removing-the-html-reports-from-fastqc" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="removing-the-html-reports-from-fastqc" class="callout-inner">
<h3 class="callout-title">Removing the HTML reports from FastQC<a class="anchor" aria-label="anchor" href="#removing-the-html-reports-from-fastqc"></a>
</h3>
<div class="callout-content">
<p>We have no use for the HTML reports produced by FastQC. Modify the
Snakefile to automatically remove them.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">Show me the solution</h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<p>Amend the <em>html</em> output of the <em>fastqc</em> rule to mark it
as <code>temporary()</code>:</p>
<pre class="source"><code>html = temporary("{indir}.{sample}_fastqc.html"),</code></pre>
<p>Since the files are already there, Snakemake will not normally remove
them unless the jobs are re-run, so you could do that as was done for
<em>trimreads</em> above. However, there is also a
<em>--delete-temp-output</em> option which forces all temporary files in
the DAG to be removed, and this provides the cleanest way to remove the
files after modifying the Snakefile.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-p</span> <span class="at">-j1</span> <span class="at">--delete-temp-output</span> multiqc</span></code></pre>
</div>
</div>
</div>
</div>
</div>
</section><section id="running-in---touch-mode"><h2 class="section-heading">Running in <em>--touch</em> mode<a class="anchor" aria-label="anchor" href="#running-in---touch-mode"></a>
</h2>
<hr class="half-width">
<p>One annoying thing about Snakemake is that, in moderately complex
workflows, it may seem determined to re-run a job for reasons that are
unclear to you. For example, after adding the <em>gzip_fastq</em> rule
above we re-ran all of the <em>kallisto</em> and <em>salmon</em>
quantifications, but with real data this could take hours or even
days.</p>
<p>Let’s pretend we just decided to <em>gzip</em> all the trimmed reads
manually, ouside of Snakemake. This results in files with new
timestamps, so to simulate this we can just <em>touch</em> all the
existing files.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> touch trimmed/<span class="pp">*</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-n</span> <span class="at">-p</span> multiqc_out</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="ex">...</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="ex">Job</span> counts:</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>    <span class="ex">count</span>   jobs</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>    <span class="ex">1</span>       multiqc</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>    <span class="ex">9</span>       salmon_quant</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>    <span class="ex">10</span></span></code></pre>
</div>
<p>Snakemake wants to re-run all the <em>salmon_quant</em> jobs (and the
<em>kallisto_quant</em> jobs, if you completed the exercise above),
which makes sense. However, we know the results are good, and don’t want
to waste time re-making them, so we can fudge the timestamps using the
<em>--touch</em> option. In the words of the Snakemake manual, “This is
used to pretend that the rules were executed, in order to fool future
invocations of snakemake.”</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> snakemake <span class="at">-j</span> 1 <span class="at">--touch</span> <span class="at">-p</span> multiqc_out</span></code></pre>
</div>
<div id="protecting-specific-files" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="protecting-specific-files" class="callout-inner">
<h3 class="callout-title">Protecting specific files<a class="anchor" aria-label="anchor" href="#protecting-specific-files"></a>
</h3>
<div class="callout-content">
<p>A related feature is the <code>protected()</code> function. This is
rather like the opposite of the <code>temporary()</code> function and
says that once an output has been produced it must not be overwritten.
In practise, Snakemake does this by revoking write permissions on the
files (as in <code>chmod -w {output}</code>).</p>
<p>This works, but can be annoying because Snakemake will refuse to run
if it believes it needs to re-create a file which is protected. An
alternative suggestion is, once you have generated an important output
file, move or copy the file away from your working directory. That way
it will be harder to accidentally clobber it (which is remarkably easy
to do!).</p>
</div>
</div>
</div>
</section><section id="shadow-mode-rules"><h2 class="section-heading">Shadow mode rules<a class="anchor" aria-label="anchor" href="#shadow-mode-rules"></a>
</h2>
<hr class="half-width">
<p>Setting your rules to run in <em>shadow</em> mode helps clean up
temporary files that are only used within the job. The <a href="https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#shadow-rules" class="external-link">Snakemake
manual describes</a> the <code>shadow</code> directive very nicely, so
refer to that for more details.</p>
<p>Briefly, in <em>shadow</em> mode, Snakemake links the input files
into a temporary working directory, then runs the <em>shell</em> command
and finally copies the outputs back. If you have ever used <a href="https://nextflow.io" class="external-link">NextFlow</a> this idea will be familiar as
NextFlow runs all workflow steps this way.</p>
<p>This test rule serves to demonstrate the operation of rules using the
<em>shadow</em> directive. The file <code>temp_file.txt</code> will not
be there after the job has run, but the <code>shadow_out.txt</code> file
will be there because Snakemake sees that it is an output file and moves
it back to the real working directory,</p>
<pre class="source"><code>rule shallow_rule:
    output: "shadow_out.txt"
    shadow: "minimal"
    shell:
        """echo minimal shadow mode
           touch shadow_out.txt
           touch temp_file.txt
           tree `pwd`
        """</code></pre>
<p>Advantages of using shadow rules are:</p>
<ul>
<li>Any temporary files created by applications do not require explicit
removal (as with <code>temp_file.txt</code> in the above example).</li>
<li>When running jobs in parallel (eg. <code>-j2</code>) certain
conflicts related to temporary files will be avoided by not ever running
multiple jobs in the same directory at once - we’ve already seen a case
back in <a href="06-awkward_programs.html">Episode 06</a> where the
final version of the <em>fastqc</em> rule has this problem.</li>
</ul>
<p>Disadvantages are:</p>
<ul>
<li>It can be confusing when error messages reference the shadow
directory.</li>
<li>Symlinks to subdirectories do not always work the way you
expect.</li>
<li>Shadow directories are not always removed cleanly if Snakemake exits
with an error.</li>
</ul>
<p>You may want to test your rules in normal mode first, then add
<code>shadow: "minimal"</code> before you run the workflow for real.</p>
<div id="removing-the-html-reports-again" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="removing-the-html-reports-again" class="callout-inner">
<h3 class="callout-title">Removing the HTML reports (again)<a class="anchor" aria-label="anchor" href="#removing-the-html-reports-again"></a>
</h3>
<div class="callout-content">
<p>Amend the <em>fastqc</em> rule once more so that the HTML reports are
not even mentioned in the rule, and will not appear in the working
directory.</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">Show me the solution</h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<p>The <code>shadow: "minimal"</code> directive will do the job nicely.
You also need to remove mention of the <code>.html</code> file from the
list of outputs and the shell commands.</p>
<pre class="code"><code>rule fastqc:
    shadow: "minimal"
    output:
        zip  = "{indir}.{myfile}_fastqc.zip"
    input:  "{indir}/{myfile}.fq"
    shell:
        """fastqc -o . {input}
           mv {wildcards.myfile}_fastqc.zip {output.zip}
        """</code></pre>
<p>In this case, marking the <em>html</em> output as
<code>temporary()</code> or simply removing the file within the
<em>shell</em> part does work fine, but the good thing about the
<em>shadow</em> approach is you do not need to deal with or mention the
unwanted file at all.</p>
</div>
</div>
</div>
</div>
<p><em>For reference, <a href="files/ep12.Snakefile">this is a
Snakefile</a> incorporating the changes made in this episode.</em></p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Cleaning up working files is good practise</li>
<li>Make use of the <code>temporary()</code> function on outputs you
don’t need to keep</li>
<li>Shadow rules can solve issues with commands that produce unwanted
files</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-13-quoting"><p>Content from <a href="13-quoting.html">Robust quoting in Snakefiles</a></p>
<hr>
<p>Last updated on 2024-03-01 |
        
        <a href="https://github.com/carpentries-incubator/snakemake-novice-bioinformatics/edit/main/episodes/13-quoting.md" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How are shell commands processed before being run?</li>
<li>How do I avoid quoting problems in filenames and commands?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Review quoting rules in the shell</li>
<li>Understand how shell command strings are processed</li>
<li>Understand how to make Snakemake commands robust</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<p><em>For reference, <a href="files/ep07.Snakefile">this is the final
Snakefile from episodes 1 to 7</a> you may use to start this
episode.</em></p>
<section id="a-review-of-quoting-rules-in-the-bash-shell"><h2 class="section-heading">A review of quoting rules in the <em>bash</em> shell<a class="anchor" aria-label="anchor" href="#a-review-of-quoting-rules-in-the-bash-shell"></a>
</h2>
<hr class="half-width">
<p>Consider the following simple <em>bash</em> shell command:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> echo Why is a <span class="st">"mouse"</span> when     it spins<span class="pp">?</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="ex">Why</span> is a mouse when it spins<span class="pp">?</span></span></code></pre>
</div>
<p>The message is printed back, but not before the shell has interpreted
the text as per various command-line parsing rules</p>
<ul>
<li>The quotes around <code>"mouse"</code> have been removed.</li>
<li>The extra spaces between <code>when</code> and <code>it</code> have
been ignored.</li>
<li>More subtly, <code>spins?</code> will be interpreted as a glob
pattern if there is any matching file:</li>
</ul>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> touch spinsX spinsY</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> echo Why is a <span class="st">"mouse"</span> when     it spins<span class="pp">?</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="ex">Why</span> is a mouse when it spinsX spinsY</span></code></pre>
</div>
<p><em>Note: if you have certain shell settings you may have seen a
warning about the unmatched glob pattern.</em></p>
<p>In shell commands, some characters are “safe”, in that <em>bash</em>
does not try to interpret them at all:</p>
<ul>
<li>Letters <code>A-Z</code> and <code>a-z</code>
</li>
<li>Digits <code>0-9</code>
</li>
<li>Period, underscore, forward slash, and hyphen <code>._/-</code>
</li>
</ul>
<p>Most other characters have some sort of special meaning and must be
enclosed in quotes if you want to pass them through verbatim to the
program you are running. This is essential with things like <em>awk</em>
commands:</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Print mean read length in FASTQ file - from https://github.com/stephenturner/oneliners</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> awk <span class="st">'NR%4==2{sum+=length($0)}END{print sum/(NR/4)}'</span> reads/ref_1_1.fq</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="ex">101</span></span></code></pre>
</div>
<p>The ‘single quotes’ ensure that the expression is passed directly to
<em>awk</em>. If “double quotes” had been used, <em>bash</em> would
replace <code>$0</code> with the name of the current script and break
the awk logic, giving a meaningless result.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># With double quotes the "$0" is substituted by *bash*, rather than being passed on to awk.</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> awk <span class="st">"NR%4==2{sum+=length(</span><span class="va">$0</span><span class="st">)}END{print sum/(NR/4)}"</span> reads/ref_1_1.fq</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="ex">0</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="co"># In an interactive shell $0 is a variable containing the value "bash"</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> echo <span class="st">"NR%4==2{sum+=length(</span><span class="va">$0</span><span class="st">)}END{print sum/(NR/4)}"</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="ex">NR%4==2{sum+=length</span><span class="er">(</span><span class="fu">bash</span><span class="kw">)</span><span class="er">}</span><span class="ex">END{print</span> sum/<span class="er">(</span><span class="ex">NR/4</span><span class="kw">)</span><span class="er">}</span></span></code></pre>
</div>
<p>So in <em>bash</em>, putting a string in single quotes is normally
enough to preserve the contents, but if you need to add literal single
quotes to the awk command for any reason you have to do some awkward
(pun intended) construction.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Yuk! But it does work...</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> awk <span class="st">'{print "\"double\" '"'"'single'"'"'"}'</span> <span class="op">&lt;&lt;&lt;</span><span class="st">''</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="st">"double"</span> <span class="st">'single'</span></span></code></pre>
</div>
</section><section id="quoting-rules-in-snakemake"><h2 class="section-heading">Quoting rules in Snakemake<a class="anchor" aria-label="anchor" href="#quoting-rules-in-snakemake"></a>
</h2>
<hr class="half-width">
<p>Anyone who has done any amount of programming or shell scripting will
have come across quoting issues in code. Bugs related to quoting can be
very troublesome. In Snakemake these are particularly complex because
the <code>shell</code> part of each rule undergoes three rounds of
interpretation before anything is actually run:</p>
<ol style="list-style-type: decimal">
<li>The string is parsed according to Python quoting rules</li>
<li>Placeholders in curly brackets (eg. <code>{input}</code>
<code>{output}</code>) are then replaced</li>
<li>The resulting string goes to the <em>bash</em> shell and is subject
to all <em>bash</em> parsing rules</li>
</ol>
<p>We’ll now look at some best practises for making your Snakefiles
robust, and some simple rules to avoid most mis-quoting
complications.</p>
<!-- The funky HTML entities in the heading below are a workaround for
     https://github.com/carpentries/sandpaper/issues/562
     I shouldn't be doing this, but in the contect of this particular episode it seems almost
     appropriate. -->
<div id="adding-a-lenreads-rule" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="adding-a-lenreads-rule" class="callout-inner">
<h3 class="callout-title">Adding a ᠎ ᠎<em>lenreads</em>᠎ ᠎ rule<a class="anchor" aria-label="anchor" href="#adding-a-lenreads-rule"></a>
</h3>
<div class="callout-content">
<p>Say we add a new rule named <em>lenreads</em>, looking very much like
the existing <em>countreads</em> rule and using the <em>awk</em>
expression we saw earlier.</p>
<pre class="source"><code>rule lenreads:
    output: "{indir}.{myfile}.fq.len"
    input:  "{indir}/{myfile}.fq"
    shell:
        "awk 'NR%4==2{sum+=length($0)}END{print sum/(NR/4)}' {input} &gt; {output}"</code></pre>
<p>Will this work as shown? If not, why not? Try it and see.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">Show me the solution</h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>It won’t work. Snakemake assumes that all parts of the string in
{curlies} are placeholders. The error will say something like
<code>NameError: The name 'sum+=length($0)' is unknown in this context</code>.
To resolve this, we have to double up all the curly braces:</p>
<pre class="source"><code>  shell:
      "awk 'NR%4==2{{"{{"}}sum+=length($0)}}END{{"{{"}}print sum/(NR/4)}}' {input} &gt; {output}"</code></pre>
</div>
</div>
</div>
</div>
</section><section id="best-practise-for-writing-robust-workflows"><h2 class="section-heading">Best practise for writing robust workflows<a class="anchor" aria-label="anchor" href="#best-practise-for-writing-robust-workflows"></a>
</h2>
<hr class="half-width">
<p>For complex commands in Snakemake rules, the <em>triple-quoted
strings</em> we saw earlier can be modified by adding a letter
<code>r</code> just before the quotes, like this.</p>
<pre class="source"><code>r"""Strings like this"""</code></pre>
<p>This “raw string” or “r-string” syntax allows embedded newlines,
literal <code>\n</code> <code>\t</code>, and both types of quotes
(<code>"</code> <code>'</code>). In other words, the interpretation as a
Python string does as little as possible, leaving most interpretation to
the <em>Bash</em> shell. This means that if you copy and paste the
commands into a shell prompt or a shell script you should get the exact
same result. You could consider using r-string quotes for all shell
commands, at the cost of a small loss of readability of the workflow
code.</p>
<p>The triple-quoting does not protect {curlies}, so if you are needing
to use <em>awk</em> commands like the one above, rather than adding
extra braces into the command you could define it as a variable.</p>
<pre class="source"><code>LEN_READS_CMD = r"""NR%4==2{sum+=length($0)}END{print sum/(NR/4)}"""

rule lenreads:
    shell:
        "awk '{LEN_READS_CMD}' {input} &gt; {output}"</code></pre>
<p>Or even better:</p>
<pre class="source"><code>rule lenreads:
    shell:
        "awk {LEN_READS_CMD:q} {input} &gt; {output}"</code></pre>
<p>Using <code>{LEN_READS_CMD:q}</code> instead of
<code>'{LEN_READS_CMD}'</code> is asking Snakemake to quote the awk
command for you. In this case, Snakemake will just put it into single
quotes, but if your variable contains single quotes or embedded newlines
or tabs or any other oddities then Snakemake will quote it robustly for
you.</p>
<p>The <code>:q</code> syntax works on any placeholder and you can
safely add it to all these placeholders:</p>
<pre class="source"><code>rule lenreads:
    shell:
        "awk {LEN_READS_CMD:q} {input:q} &gt; {output:q}"</code></pre>
<p>Now the <em>lenreads</em> rule would be able to work on an input file
that contains spaces or other unusual characters. Also, if the
<em>input</em> is a list of files, this will still work just fine,
whereas <code>'{input}'</code> will fail as it just combines all the
filenames into one big string within the quotes.</p>
<p>In general, choose file names that only contain shell-safe characters
and no spaces, but if you can’t do that then in most cases ensuring all
your placeholders have <code>:q</code> will be enough to keep things
working.</p>
<div id="handling-filenames-with-spaces" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="handling-filenames-with-spaces" class="callout-inner">
<h3 class="callout-title">Handling filenames with spaces<a class="anchor" aria-label="anchor" href="#handling-filenames-with-spaces"></a>
</h3>
<div class="callout-content">
<p>Use the following command to rename the <em>temp33</em> and
<em>etoh60</em> samples so that the filenames contain spaces:</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">BASH<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="sourceCode bash" tabindex="0"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> rename <span class="at">-v</span> <span class="at">-s</span> <span class="st">'etoh'</span> <span class="st">'etoh '</span> <span class="at">-s</span> <span class="st">'temp'</span> <span class="st">'temp '</span> reads/<span class="pp">*</span>.fq</span></code></pre>
</div>
<p>Fix the workflow so you can still run the MultiQC report over all the
samples (run Snakemake with <code>-F</code> to check that all the steps
really work).</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">Show me the solution</h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<p>This just involves adding <code>:q</code> to a whole bunch of
placeholders. Unless you are very diligent it will probably take a few
tries to catch every one of them and make the workflow functional
again.</p>
</div>
</div>
</div>
</div>
<div id="external-scripts" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="external-scripts" class="callout-inner">
<h3 class="callout-title">External scripts<a class="anchor" aria-label="anchor" href="#external-scripts"></a>
</h3>
<div class="callout-content">
<p>If rules in your Snakefile end up containing a large amount of shell
code, or you are running into multiple quoting issues, this is probably
a sign you should move the code to a separate Bash script. This will
also make it easier to test the code directly.</p>
<p>To make integration with these Bash scripts easier, Snakemake has a
<a href="https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#bash" class="external-link">script
field</a> which you can put as an alternative to the <code>shell</code>
field. When scripts are run like this, Snakemake passes the parameters
directly into the script as associative arrays. The Snakemake manual has
a good expalantion, with examples, of how this works.</p>
</div>
</div>
</div>
<p><em>For reference, <a href="files/ep13.Snakefile">this is a
Snakefile</a> incorporating the changes made in this episode.</em></p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Having a grasp of string quoting rules is boring but important</li>
<li>Understand the three processing steps each <em>shell</em> command
goes through before it is actually run</li>
<li>Put complex shell commands into raw triple-quotes
(<code>r"""</code>)</li>
<li>Watch out for commands that have {curly brackets}, and double them
up</li>
<li>Use the built-in <code>:q</code> feature to protect arguments from
<em>bash</em> interpretation</li>
</ul>
</div>
</div>
</div>
</section></section>
</div>
    </main>
</div>
<!-- END  : inst/pkgdown/templates/content-extra.html -->

      </div>
<!--/div.row-->
      		<footer class="row footer mx-md-3"><hr>
<div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>
        
        <a href="https://github.com/carpentries-incubator/snakemake-novice-bioinformatics/edit/main/README.md" class="external-link">Edit on GitHub</a>
        
	
        | <a href="https://github.com/carpentries-incubator/snakemake-novice-bioinformatics/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries-incubator/snakemake-novice-bioinformatics/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-incubator/snakemake-novice-bioinformatics/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:tim.booth@ed.ac.uk">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">
        
        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>
        
        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.4" class="external-link">sandpaper (0.16.4)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.5" class="external-link">pegboard (0.7.5)</a>, and <a href="https://github.com/carpentries/varnish/tree/1.0.2" class="external-link">varnish (1.0.2)</a></p>
			</div>
		</footer>
</div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://carpentries-incubator.github.io/snakemake-novice-bioinformatics/aio.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "All in One View",
  "creativeWorkStatus": "active",
  "url": "https://carpentries-incubator.github.io/snakemake-novice-bioinformatics/aio.html",
  "identifier": "https://carpentries-incubator.github.io/snakemake-novice-bioinformatics/aio.html",
  "dateCreated": "2021-03-26",
  "dateModified": "2024-04-16",
  "datePublished": "2024-04-16"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code -->
</body>
</html><!-- END:   inst/pkgdown/templates/layout.html-->

